---
title: "Final clean analysis of long-term infection RNA-seq"
output: 
  html_notebook: 
    fig_width: 12
    fig_height: 10
editor_options: 
  chunk_output_type: console
---

```{r include=TRUE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(include = TRUE)
library(tidyverse)
library(greekLetters)
theme_set(theme_bw(base_size = 27))
```

```{r}
counts <- read.table("all_reads_STAR", header = F, sep = "\t", stringsAsFactors = F)
#Next manipulate the table to remove redundant columns and only keep relevant columns (1 columns with gene names, and the rest with counts for reverse stranded library)
counts_simple <- counts %>% 
  dplyr::select(Gene_ID = V1, UI92 = V4, UI93 = V8, UI96 = V12, N92 = V16, N93 = V20, WT96 = V24, WT91 = V28, WT94 = V32, N95 = V36)
filter <- apply(counts_simple [,2:10],1,function(x) mean(x)>5)
counts_simple <- counts_simple [filter,]
```

```{r vis-align-stats}
#Visualize alignment statistics
counts_stats <- counts_simple %>% 
  dplyr::slice(1:4) %>% 
  pivot_longer(-Gene_ID, values_to = "numbers", names_to = "mapping")
mapped <-  counts_simple %>%
  select(-Gene_ID) %>%
  summarise(across(everything(), sum)) %>% 
  pivot_longer(everything(),values_to = "numbers", names_to = "mapping") %>% 
  mutate(Gene_ID = "Mapped") %>% 
  bind_rows(counts_stats) %>% 
  mutate(Gene_ID = fct_relevel(str_to_title(str_replace(Gene_ID, "N_","")), "Mapped"))

#Raw counts stats
raw_count_plot <- ggplot(mapped, aes(x=mapping, y=numbers, fill = Gene_ID)) +
  geom_bar(stat = "identity") + scale_fill_brewer(palette = "Accent") +
  labs(x = "", y = "Raw counts") +theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6)) + 
  guides(fill = guide_legend(title = "Mapping Status"))
calc_relative <- function(x) x/sum(x) * 100
#Relative count stats
rel_map <- mapped %>% 
  pivot_wider(Gene_ID, names_from = mapping, values_from = numbers) %>%  
  mutate(across(-Gene_ID, calc_relative)) %>% 
  pivot_longer(-Gene_ID,values_to = "numbers", names_to = "mapping")
rel_count_plot <-ggplot(rel_map, aes(x=mapping, y=numbers, fill = Gene_ID)) +
  geom_bar(stat = "identity") + scale_fill_brewer(palette = "Accent") +
  labs(x = "", y = "Proportion of counts") +
  theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6)) + 
  guides(fill = guide_legend(title = "Mapping Status"))
#pdf(file = here::here("final clean figures/mapping stats.pdf"), width = 12, height = 10)
raw_count_plot
rel_count_plot
#dev.off()
```

Calculate statistics of types of transcripts mapped

```{r}
library(org.Hs.eg.db)
geneids <- mapIds(org.Hs.eg.db, keys = str_replace_all(counts_simple$Gene_ID [-1:-4], "\\.[1-9]([0-9])?", ""), 
                  keytype = "ENSEMBL", column = "GENETYPE", multiVals = "first") %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "Gene_ID")
c_type <- counts_simple %>% 
  dplyr::slice(-1:-4) %>% 
  mutate(Gene_ID = str_replace_all(Gene_ID, "\\.[1-9]([0-9])?", "")) %>% 
  left_join(geneids) %>% 
  dplyr::select(-Gene_ID) %>%
  rename_with(function(x) "Type",last_col()) %>% 
  group_by(Type) %>% 
  summarise(across(everything(),sum)) %>% 
  mutate(across(-Type, calc_relative)) %>% 
  pivot_longer(-Type, values_to = "numbers", names_to = "mapping") %>% 
  mutate(Type = str_replace_na(Type, replacement = "other"))
rel_type_plot <- ggplot(c_type, aes(x=mapping, y=numbers, fill = fct_reorder(Type,numbers, .desc = T))) +
  geom_bar(stat = "identity") + scale_fill_brewer(palette = "Accent") +
  labs(x = "", y = "Proportion of genes") +
  theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6)) + 
  guides(fill = guide_legend(title = "Gene type"))
#pdf(file = here::here("final clean figures/mapping stats by gene type.pdf"), width = 12, height = 10)
rel_type_plot
#dev.off()
```

Apply EDAseq normalization to the data

```{r edaseq}
#Use EDAseq to normalize the libraries
library(EDASeq)
c_mat <- counts_simple %>% 
                     dplyr::select(-where(is.character)) %>% 
                     dplyr::slice(-1:-4) %>% 
                     as.matrix()
rownames(c_mat) <- counts_simple$Gene_ID [-1:-4]
pData <- data.frame(genotype=vector(mode = "logical", length = ncol(c_mat)),
                    timepoint = vector(mode = "logical", length = ncol(c_mat)),
                    samples = colnames(c_mat),
                    row.names = colnames(c_mat))
pData <- pData %>% 
  mutate(genotype = case_when(
    str_detect(samples, "WT") ~ "WT",
    str_detect(samples, "UI") ~ "UI",
    str_detect(samples, "N") ~ "N"
  ), timepoint = 9) %>%
  dplyr::select(-samples)

data <- newSeqExpressionSet(counts=c_mat,
                            phenoData=pData)
#Normalize the data next
dataNorm <- betweenLaneNormalization(data, which="full") 
#Plot the counts 
#pdf(here::here("final clean figures/Normalization effect on the library.pdf"), width = 12, height = 10)
boxplot(data, main = "Unnormalized counts", col = "dodgerblue3")
boxplot(dataNorm, main = "Normalized counts", col = "dodgerblue3")
#dev.off()
#For DE we will create this analysis with offset values
dataOffset <- betweenLaneNormalization(data,
                                       which="full",offset=TRUE)
```


Next step is to use normalized counts in the DESEq2 analysis of differential expression
```{r deseq}
library(DESeq2)
#Prepare the DESEq2 object 
pData$group <- factor(paste0(pData$genotype,pData$timepoint))
dds <- DESeqDataSetFromMatrix(countData = c_mat,
                              colData = pData,
                              design = ~ group)
dds$group <- relevel(dds$group, ref= "UI9")

#Perform the DESEq2 analysis
dds <- DESeq(dds)
rld <- rlog(dds)

normal_PCA <- plotPCA(rld, returnData = T, intgroup = "timepoint")
#Plot the PCA graphs of the normalized counts
#png(here::here("final clean figures/PCA plot of unnormalized counts.png"), width = 12, height = 10,
 #  units = "in", res = 720)
ggplot(data = normal_PCA, aes(shape = pData$genotype, color = pData$genotype)) +
  scale_color_manual(name = "Infection status",
                    labels = c("N" = paste0(greeks("Delta"),"EscN EPEC"),
                                "WT" = "Wild-type EPEC", "UI" = "Uninfected"),
                     values = c("UI" = "coral","WT" = "darkblue", "N" = "forestgreen")) +
  geom_point(mapping = aes(x = PC1, y = PC2), size = 4) +
  labs(title = "Raw DESeq2 PCA plot") +
  scale_shape_manual(name = "Infection status",
                     labels = c("N" = paste0(greeks("Delta"),"EscN EPEC"),
                                "WT" = "Wild-type EPEC", "UI" = "Uninfected"),
                     values = c("UI" = 15, "WT" = 16, "N" = 17))
#dev.off()
```

Identify differentially expressed genes
```{r deseq-res}
resWTvsUI <- results(dds, contrast = c("group","WT9", "UI9"))
resWTvsN <- results(dds, contrast = c("group","WT9", "N9"))
resNvsUI <- results(dds, contrast = c("group","N9", "UI9"))
res <- list(resWTvsUI,resWTvsN,resNvsUI) %>% 
  map(as.data.frame) %>% 
  map(rownames_to_column, var = "rowname") %>% 
  purrr::reduce(left_join, by = "rowname") %>% 
  rename_with(function(x) str_replace(x,"$",".NvsUI"),
              .cols = !matches("\\.[xy]")) %>% 
  rename_with(function(x) ifelse(str_detect(x,"\\.x"),
                                 str_replace(x,"\\.x",".WTvsUI"),
                                 str_replace(x,"\\.y",".WTvsN")),
              .cols = matches("\\.[xy]")) %>% 
  rename_with(function(x) "Gene ID",1)
map(list(resWTvsUI,resWTvsN,resNvsUI),summary)
#write.table(res, here::here("final clean figures/unnormalized results.txt"),
 #           sep = "\t", row.names = F)
```


Analyze loadings of the principal components in the data
```{r pcatools}
library(PCAtools)
library('org.Hs.eg.db')
#Create a mapping for ID to symbol
geneids <- mapIds(org.Hs.eg.db, keys = str_replace_all(rownames(counts(dds)), "\\.[1-9]([0-9])?", ""), 
                  keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")

cnts <- assay(rld)
#Confirm that that the mapping is correct
all(str_replace_all(rownames(counts(dds)), "\\.[1-9]([0-9])?", "") == names(geneids))
#Rename rows that had a symbol name
newnames <- ifelse(is.na(geneids) | duplicated(geneids),
    names(geneids), geneids)
rownames(cnts) <- newnames
p <- pca(cnts,colData(dds), removeVar = 0.1)
p$metadata$genotype <- p$metadata$genotype %>% as.factor() %>% fct_relevel("UI")
#pdf(here::here("final clean figures/PCA tools analysis of PCA.pdf"), width = 12, height = 10)
biplot(p, colby = "genotype", shape = "genotype", showLoadings = T,gridlines.major = FALSE)
biplot(p, colby = "genotype", shape = "genotype",gridlines.major = FALSE)
plotloadings(p,  rangeRetain = 0.01,
    labSize = 4.0,
    caption = "Top 1% variables")
eigencorplot(p,
             components = getComponents(p,c(1:9)),
             metavars = c('group','genotype'))
eigencorplot(p,
             components = getComponents(p,1:9),
    metavars = c('group','genotype'),
    col = c('white', 'cornsilk1', 'gold', 'forestgreen', 'darkgreen'),
    cexCorval = 1.2,
    fontCorval = 2,
    posLab = 'all',
    rotLabX = 45,
    scale = TRUE,
    main = bquote(Principal ~ component ~ Pearson ~ r^2 ~ correlates),
    plotRsquared = TRUE,
    corFUN = 'pearson',
    corUSE = 'pairwise.complete.obs',
    corMultipleTestCorrection = 'BH',
    signifSymbols = c('****', '***', '**', '*', ''),
    signifCutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1))
eigencorplot(p,
             components = getComponents(p,1:9),
    metavars = c('group','genotype'),
    col = c('white', 'cornsilk1', 'gold', 'forestgreen', 'darkgreen'),
    cexCorval = 1.2,
    fontCorval = 2,
    posLab = 'all',
    rotLabX = 45,
    scale = TRUE,
    main = bquote(Principal ~ component ~ Pearson ~ r^2 ~ correlates ~ no ~ correction),
    plotRsquared = TRUE,
    corFUN = 'pearson',
    corUSE = 'pairwise.complete.obs',
    corMultipleTestCorrection = 'none',
    signifSymbols = c('****', '***', '**', '*', ''),
    signifCutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1))
#dev.off()
```
Make volcano plots of differentially expressed genes
```{r volcano}
library(EnhancedVolcano)
library('org.Hs.eg.db')
#shrink results
resWTUIlfc <- lfcShrink(dds, coef="group_WT9_vs_UI9", type="apeglm", res = resWTvsUI)
resNUIlfc <- lfcShrink(dds, coef="group_N9_vs_UI9", type="apeglm", res = resNvsUI)
#Get results for WTvs EscN
dds.N <- dds
dds.N$group <- relevel(dds.N$group, ref= "N9")
dds.N <- DESeq(dds.N)
resWTvsN <- results(dds.N, contrast = c("group","WT9", "N9"))
resWTNlfc <- lfcShrink(dds.N, coef="group_WT9_vs_N9", type="apeglm", res = resWTvsN)
#Create a mapping for ID to symbol
geneids <- mapIds(org.Hs.eg.db, keys = str_replace_all(rownames(resWTNlfc), "\\.[1-9]([0-9])?", ""), 
                  keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first") %>% 
  as.data.frame() %>% 
  bind_cols(rownames(resWTNlfc)) %>% 
  dplyr::select("symbol" = 1, "rowname" = `...2`)
#Get nice names or keep ensembl IDs
naOrDup <- is.na(geneids$symbol) | duplicated(geneids$symbol)
geneids$gene_name <- ifelse(naOrDup, rownames(geneids), geneids$symbol)
#identify top20 genes for each comparison
id_top <- function(x) {
  x %>% 
  as.data.frame() %>% 
  arrange(padj) %>% 
  dplyr::slice(1:20) %>% 
  rownames_to_column(var = "rowname") %>% 
  left_join(geneids) }
top_list <- list("WTN" = resWTNlfc,"WTUI" = resWTUIlfc, "NUI" = resNUIlfc) %>% 
  map(id_top)
results_list <- list(resWTNlfc,resWTUIlfc,resNUIlfc)
#Replace the rownames with gene names
for (i in 1:3) {rownames(results_list [[i]]) <- geneids$gene_name}
rownames(resWTNlfc) <- geneids$gene_name
#pdf(here::here("final clean figures/Volcano plots.pdf"), width = 12, height = 10)
plotMA(resWTvsN, main = paste0("WT infection vs ", greeks("Delta"),  "EscN infection\nResults without shrinking"))
plotMA(resWTNlfc, main = paste0("WT infection vs ", greeks("Delta"),  "EscN infection\nResults after apeglm shrinking"))
plotMA(resWTvsUI, main = "WT infection vs uninfected cells \nResults without shrinking")
plotMA(resWTUIlfc, main = "WT infection vs uninfected cells \nResults after apeglm shrinking")

#Create a wrapper function for enhancedvolcano to plot all the graphs
volcano_with_labels <- function(res_obj,select_lab = "",title) {
  EnhancedVolcano(res_obj,
    lab = rownames(res_obj),
    selectLab = select_lab,
    x = 'log2FoldChange',
    y = 'pvalue',
    title = title,
    pointSize = 3,
    col = c("darkgrey","red3","dodgerblue2","forestgreen"),
    legendPosition = 'right',
    subtitle = NULL,
    pCutoff = 0.05,
    FCcutoff = 0.6,
    labSize = 4.0,
                boxedLabels = T, labCol = 'black',
               labFace = 'bold',
                drawConnectors = T, widthConnectors = 1.0,
               colConnectors = 'black')
}
titles <- list(paste0("Differentially expressed genes between WT and ",greeks("Delta"),"EscN infected cells"),
               paste0("Differentially expressed genes between WT and UI cells"),
               paste0("Differentially expressed genes between",greeks("Delta"),"EscN infected and uninfected cells"))
for (i in 1:3) {
  volcano_with_labels(results_list[[i]],title = titles [[i]]) %>% print()
  volcano_with_labels(results_list[[i]],select_lab = top_list [[i]]$gene_name,title = titles [[i]]) %>% print()
}
#dev.off()
```
Apply EGSEA to differentially expressed data

```{r eggsea-prep}
library(limma)
#remove intercept since this is a factor. Intercept or no intercept does not make a difference
#That is what the 0+ means in the formula
design_matrix <- model.matrix(~0+pData$group) 
colnames(design_matrix) <- colnames(design_matrix) %>% str_replace("pData\\$group","") 
names_for_contr <- resultsNames(dds) [-1] %>%
  str_replace_all("group_","") %>% 
  str_replace_all("_vs_","-")
names_for_contr [3] <- "WT9-N9"
contrast_obj <- makeContrasts(contrasts = names_for_contr, levels = colnames(design_matrix))
group_obj <- as.factor(pData$group)
cnt_obj <- dataNorm@assayData$normalizedCounts
#finally prepare the geneID mapping tool
library('org.Hs.eg.db')
geneids <- mapIds(org.Hs.eg.db, keys = str_replace(rownames(cnt_obj), "\\.[1-9]([0-9])?", ""), keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first")

#Replace the rownames of the count matrix with ENTREZ ID
rownames(cnt_obj) <- unname(geneids)
#Remove NAs from the matrix (not every gene had an NCBI match)
cnt_obj_filt <- cnt_obj [!is.na(rownames(cnt_obj)),] 
#Deal with the remaining duplicates
dupes <-  c(which(duplicated(rownames(cnt_obj_filt)) | duplicated(rownames(cnt_obj_filt),fromLast = T)))
means <- data.frame(matrix(nrow = length(dupes))) %>% mutate(index = dupes)
#Select the duplicate with higher gene count
for (i in 1:length(dupes)) {
  means$avg [i] <- mean(cnt_obj_filt[dupes[i],])
  means$entrez [i] <- rownames(cnt_obj_filt) [dupes [i]]
}
means$toremove <- "FALSE"
i <- 1
c <- 1
while (i <= length(dupes)) {
  if (i==c) {
    c <- c+1
  } else if (means$entrez [i] == means$entrez [c]) {
    if(means$avg [i] > means$avg [c]) {
      means$toremove [c] <- T
    } else {
      means$toremove [i] <- T
    }
    i <- i+1
    c <- 1
  } else {
    c <- c+1
  }
}
dupes_to_remove <- means %>% filter(toremove == T)
cnt_obj_filt <- cnt_obj_filt [-dupes_to_remove$index,]
geneids_smbl <- mapIds(org.Hs.eg.db, keys = rownames(cnt_obj_filt), keytype = "ENTREZID", column = "SYMBOL", multiVals = "first") %>% 
  unlist() 
geneids_smbl <- data.frame(entrezid = names(geneids_smbl), symbol = unname(geneids_smbl))
```


```{r eggsea, eval=FALSE}
library(EGSEA)
#Move onto EGSEA analysis
#Build gene set collection
gs.annots <-  buildIdx(entrezIDs = rownames(cnt_obj_filt), species = "human",
msigdb.gsets = "none")


#Perform the EGSEA analysis with voom built-in
gsa <-  egsea.cnt(counts = cnt_obj_filt, group = group_obj, design = design_matrix,
  contrasts = contrast_obj, gs.annots = gs.annots, symbolsMap = geneids_smbl,
 baseGSEAs = egsea.base(), sort.by = "p.adj", fdr.cutoff = 0.1, combineMethod = "wilkinson",
  num.threads = 4, report = TRUE, report.dir = here::here("final clean figures/all comparisons"),
  verbose = T, display.top = 35)
```

Write all the necessary objects as text tables to combine with miRNA results
```{r write-tables, eval=FALSE}
write.table(cnt_obj_filt, here::here("final clean figures/count_matrix_for_egsea.txt"),
            sep = "\t", row.names = T)
write.table(counts(dds), here::here("final clean figures/count_matrix_from_dds.txt"),
            sep = "\t", row.names = T)
write.table(normalizationFactors(dds), here::here("final clean figures/norm_factors_from_dds.txt"),
            sep = "\t", row.names = T)
write_tsv(geneids_smbl, here::here("final clean figures/egsea_symbol_mapping.txt"))
```
