---
title: "finalized analysis of EPEC data"
output: 
  html_notebook:
    fig_width: 6
    fig_height: 5
editor_options: 
  chunk_output_type: inline
---


```{r lib-setup, results='hide', message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(include = TRUE)
knitr::opts_chunk$set(fig.width = 6)
knitr::opts_chunk$set(fig.height = 5)
library(greekLetters)
library(RUVSeq)
library(EDASeq)
library(DESeq2)
library(PCAtools)
library('org.Hs.eg.db')
library(EGSEA)
library(limma)
library(pheatmap)
library(eulerr)
library(UpSetR)
library(clusterProfiler)
library(tidyverse)
library(ggprism)
library(rcartocolor)
library(patchwork)
```

Define all helper functions and color choices
```{r helper, include=TRUE}
theme_set(theme_prism(base_size = 14) + theme(legend.title = element_text()))
#Set color choices from CARTOcolors
discrete_pal_choice <- "Bold"
cont_pal_choice <- "BluYl"
diverg_pal_choice <- "Geyser"
#Set color vectors
cont_colors <- carto_pal(name = cont_pal_choice)
chosen_cont_colors <- cont_colors [c(2,4,6,7)]
#Helper functions
calc_relative <- function(x) x/sum(x) * 100
#Set up paths for data, and results
data_path <- "short_infection_EPEC_reads/data/"
res_path <- "short_infection_EPEC_reads/results/"
```

#Read in the count data and rename the columns to fit the sample names

```{r readorigstar}
counts <- read.table(here::here(str_c(data_path,"STAR_EPEC_reads_updated.txt")),header = F, sep = "\t", stringsAsFactors = F)
#Next manipulate the table to remove redundant columns and only keep relevant columns (1 columns with gene names, and the rest with counts for reverse stranded library
counts_simple <- counts [,seq(from=4,to=96,by=4)]
colnames(counts_simple) <- c("4UIN", "4UIWT", "490WT", "490N", "4180N", "4180WT", "5UIN", "5UIWT", "545WT", "545N", "590WT", "590N", "5180N", "5180WT", "6UIN", "6UIWT", "645WT", "645N", "690WT", "690N", "6180N", "6180WT", "445N", "445WT")
counts_simple <- counts_simple %>% 
  dplyr::select(-`6UIN`)
counts_simple$Gene_ID <- counts$V1

filter <- apply(dplyr::select(counts_simple,-Gene_ID),1,function(x) mean(x)>5)
counts_simple <- counts_simple [filter,] %>% 
  mutate(Gene_ID = str_remove(Gene_ID, "(gene:)|(transcript:)"))
  
#write.table(counts_simple,file=here::here(str_c(res_path,"tables/only_reverse_strand_counts_cleaned.txt")),sep = "\t", row.names = F)
```

Visualize EscN counts per sample:
The EscN gene is: E2348C_3948
```{r escn-counts, eval=FALSE, fig.width=6, fig.height=8}
escn_counts <- counts_simple %>% 
  filter(Gene_ID == "E2348C_3948") %>% 
  pivot_longer(-Gene_ID, names_to = "samples", values_to = "numbers") %>% 
  mutate(time = if_else(str_detect(samples,"UI"), 
                        0, 
                        as.numeric(str_remove(str_extract(samples, "[:digit:]+"), "^[:digit:]"))),
         strain = if_else(str_detect(samples,"WT"), str_c("bold(WT)"),str_c("bold(Delta)*bold(EscN)"))) %>% 
  ggplot() + 
  geom_point(aes(x=fct_reorder(samples,time), y = numbers), size = 3) +
  labs(y = "counts", title = "counts for EscN gene", x = element_blank()) +
  facet_wrap(~strain, ncol = 1, scales = "free", labeller = label_parsed) +
  theme(axis.text.x = element_text(angle = 75, vjust = 0.6))
escn_counts_toscale <- escn_counts + facet_wrap(~strain, scales = "free_x" ,ncol = 1, labeller = label_parsed)

#ggsave(here::here(str_c(res_path,"raw counts for escN.png")),plot = escn_counts,width = 6, height = 8, dpi = 1080)
#ggsave(here::here(str_c(res_path,"raw counts for escN_toscale.png")),plot = escn_counts_toscale, width = 6, height = 8, dpi = 1080)
```


Visualize mapping stats
```{r vis-align-stats, fig.width=8, fig.height=5.5}
#Visualize alignment statistics
counts_stats <- counts_simple %>% 
  dplyr::slice(1:4) %>% 
  pivot_longer(-Gene_ID, values_to = "numbers", names_to = "mapping")
mapped <-  counts_simple %>%
  select(-Gene_ID) %>%
  summarise(across(everything(), sum)) %>% 
  pivot_longer(everything(),values_to = "numbers", names_to = "mapping") %>% 
  mutate(Gene_ID = "Mapped") %>% 
  bind_rows(counts_stats) %>% 
  group_by(mapping) %>% 
  mutate(rel_numbers = calc_relative(numbers)) %>% 
  mutate(Gene_ID = fct_relevel(str_to_title(str_replace(Gene_ID, "N_","")), "Mapped")) %>% 
  mutate(time = if_else(str_detect(mapping,"UI"), 
                        0, 
                        as.numeric(str_remove(str_extract(mapping, "[:digit:]+"), "^[:digit:]"))),
         strain = if_else(str_detect(mapping,"WT"), str_c("bold(WT)"),str_c("bold(Delta)*bold(EscN)")))

#Raw counts stats
raw_count_plot <- ggplot(mapped, aes(x=fct_reorder(mapping,time), y=numbers, fill = Gene_ID)) +
  geom_bar(stat = "identity") + scale_fill_carto_d(palette = discrete_pal_choice) +
  labs(x = "", y = "Raw counts") +
  theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6)) + 
  guides(fill = guide_legend(title = "Mapping Status")) +
  facet_wrap(~strain, scales = "free", labeller = label_parsed)
#Relative count stats
rel_count_plot <- ggplot(mapped, aes(x=fct_reorder(mapping,time), y=rel_numbers, fill = Gene_ID)) +
  geom_bar(stat = "identity") + scale_fill_carto_d(palette = discrete_pal_choice) +
  labs(x = "", y = "Proportion of counts") +
  theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6),
        legend.title = element_text()) + 
  guides(fill = guide_legend(title = "Mapping Status")) +
  facet_wrap(~strain, scales = "free", labeller = label_parsed)

#pdf(file = here::here(str_c(res_path,"mapping stats.pdf")), width = 8, height = 5.5)
raw_count_plot
rel_count_plot
#dev.off()
```

Calculate statistics of types of transcripts mapped

```{r map_by_type, fig.width=8, fig.height=5}
#Prepare gene annotation from the gff3 file
geneids <- read_delim(here::here(str_c(data_path,"EPEC_annot.rel-41.gff3")), 
       delim = "\t", escape_double = FALSE, 
       col_names = FALSE, trim_ws = TRUE, skip = 11) 
annot <- geneids %>% 
  filter(X3 == "gene" | X3 == "pseudogene" | X3 == "rRNA_gene" | 
           X3 == "tRNA_gene" | X3 == "ncRNA_gene") %>% 
  filter(str_detect(X9,"ID=gene")) %>% 
  select(X9) %>% 
  separate(X9, into = c("ID", "Name", "Biotype", "Description", "Geneid"), extra = "drop", sep = ";") %>% 
  mutate(ID = str_remove(ID, "ID=gene:"),
         true_name = case_when(
           str_detect(Name, "Name") ~ str_remove(Name, "Name="),
           str_detect(Name, "biotype") ~ ID
           ), true_biotype = case_when(
             str_detect(Name, "biotype") ~ str_remove(Name, "biotype="),
             str_detect(Biotype, "biotype") ~ str_remove(Biotype, "biotype=")
           )) %>% 
  mutate(true_name = if_else(duplicated(true_name), ID, true_name))
c_type <- counts_simple %>% 
  dplyr::slice(-1:-4) %>%
  left_join(dplyr::select(annot, ID, true_biotype), by = c("Gene_ID" = "ID")) %>% 
  dplyr::select(-Gene_ID) %>%
  rename_with(function(x) "Type",last_col()) %>% 
  group_by(Type) %>% 
  summarise(across(everything(),sum)) %>% 
  reshape2::melt() %>% 
  group_by(variable) %>% 
  mutate(rel_numbers = calc_relative(value)) %>% 
  mutate(time = if_else(str_detect(variable,"UI"), 
                        0, 
                        as.numeric(str_remove(str_extract(variable, "[:digit:]+"), "^[:digit:]"))),
         strain = if_else(str_detect(variable,"WT"), str_c("bold(WT)"),str_c("bold(Delta)*bold(EscN)")))
c_type_no_mrna <- c_type %>% 
  filter(Type != "protein_coding") %>% 
   group_by(variable) %>% 
  mutate(rel_numbers = calc_relative(value))

rel_type_plot <- ggplot(c_type, aes(x=fct_reorder(variable,time), y=rel_numbers, fill = fct_reorder(Type,rel_numbers, .desc = T))) +
  geom_bar(stat = "identity") + scale_fill_carto_d(palette = discrete_pal_choice) +
  labs(x = "", y = "Proportion of genes", fill = "Gene type") +
  theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6)) +
  facet_wrap(~strain, scales = "free", labeller = label_parsed)
rel_type_plot_no_mrna <- ggplot(c_type_no_mrna, 
                                aes(x=fct_reorder(variable,time), y=rel_numbers, fill = fct_reorder(Type,rel_numbers, .desc = T))) +
  geom_bar(stat = "identity") + scale_fill_carto_d(palette = discrete_pal_choice) +
  labs(x = "", y = "Proportion of genes", fill = "Gene type") +
  theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6)) +
  facet_wrap(~strain, scales = "free", labeller = label_parsed)
#pdf(file = here::here(str_c(res_path,"mapping stats by gene type.pdf")), width = 8, height = 5)
rel_type_plot 
rel_type_plot_no_mrna
#dev.off()
```


Apply EDAseq normalization to the data

```{r edaseq}
#Use EDAseq to normalize the libraries
c_mat <- counts_simple %>% 
                     dplyr::select(-dplyr::last_col()) %>% 
                     dplyr::slice(-1:-4) %>% 
                     as.matrix()
rownames(c_mat) <- counts_simple$Gene_ID [-1:-4]
pData <- data.frame(genotype=vector(mode = "logical", length = ncol(c_mat)),
                    timepoint = vector(mode = "logical", length = ncol(c_mat)),
                    samples = colnames(c_mat),
                    row.names = colnames(c_mat))
#Here we are setting all uninfected samples as WT to pool 0hr timepoints
pData <- pData %>% 
  mutate(genotype = case_when(
    str_detect(samples, "WT") ~ "WT",
    str_detect(samples, "N") ~ "N"
  ), timepoint = if_else(str_detect(samples,"UI"), 
                        0, 
                        as.numeric(str_remove(str_extract(samples, "[:digit:]+"), "^[:digit:]")))) %>% 
  dplyr::select(-samples)

data <- newSeqExpressionSet(counts=c_mat,
                            phenoData=pData)
#Normalize the data next
dataNorm <- betweenLaneNormalization(data, which="full") 
#Plot the counts 
#pdf(here::here(str_c(res_path,"Normalization effect on the library.pdf")), width = 8, height = 6.3)
boxplot(data, main = "Unnormalized counts", col = "dodgerblue3")
boxplot(dataNorm, main = "Normalized counts", col = "dodgerblue3")
#dev.off()
#For DE we will create this analysis with offset values
dataOffset <- betweenLaneNormalization(data,
                                       which="full",offset=TRUE)
```
Next step is to use normalized counts in the DESEq2 analysis of differential expression
```{r deseq}
#Prepare the DESEq2 object and apply EDASeq normalization factors
pData <- pData %>% 
  rownames_to_column(var = "samples") %>% 
  mutate(genotype = if_else(timepoint ==0, str_remove(samples, "^.{3,3}"), genotype)) %>% 
  column_to_rownames(var = "samples")
pData$group <- factor(paste0(pData$genotype,pData$timepoint))
dds <- DESeqDataSetFromMatrix(countData = counts(data),
                              colData = pData,
                              design = ~ group)
dds$group <- relevel(dds$group, ref= "WT0")
normFactors <- exp(-1 * offst(dataOffset))
normFactors <- normFactors / exp(rowMeans(log(normFactors)))
normalizationFactors(dds) <- normFactors

#Perform the DESEq2 analysis
dds <- DESeq(dds)
rld <- rlog(dds)

normal_PCA <- plotPCA(rld, returnData = T, intgroup = "group") %>% 
  mutate(timepoint = as.numeric(str_remove(group, "[:alpha:]+")),
         genotype = ifelse(timepoint == 0, str_remove(name, "[:digit:]+UI"), str_remove(name, "[:digit:]+"))) %>% 
  mutate(genotype = fct_relevel(genotype, "WT", "N"))
  
#Plot the PCA graphs of the normalized counts
png(here::here(str_c(res_path,"PCA plot of normalized counts.png")), width = 7, height = 5, units = "in", res = 1080)
ggplot(data = normal_PCA, aes(shape = genotype)) +
   scale_color_manual(values = chosen_cont_colors,breaks = c(0,45,90,180)) +
  #scale_color_carto_c(palette = cont_pal_choice, breaks = c(0,45,90,180)) +
  geom_point(mapping = aes(x = PC1, y = PC2, color = factor(timepoint)), size = 5) +
  labs(title = "EDAseq normalized DESeq2 PCA plot \nnot including the outlier",
       x = "PC1: 40% variance",
       y = "PC2: 14% variance") +
  guides(color = guide_legend(title = "Time of infection (min)"),
         shape = guide_legend(title = "Strain")) +
  scale_shape_manual(labels = c("WT" = "Wild-type EPEC",
                                "N" = paste0(greeks("Delta"),"EscN EPEC")),
                     values = c("Uninfected" = 15, "WT" = 16, "N" = 17))
dev.off()
```
Identify differentially expressed genes
```{r deseq-res}
res45 <- results(dds, contrast = c("group","WT45", "WT0"))
res90 <- results(dds, contrast = c("group","WT90", "WT0"))
res180 <- results(dds, contrast = c("group","WT180", "WT0"))
time_res <- list(res45,res90,res180) %>% 
  map(as.data.frame) %>% 
  map2(c(".45",".90",".180"), ~rename_with(.x, .fn = function(colname) str_c(colname,.y))) %>% 
  map(rownames_to_column, var = "geneid") %>% 
  reduce(left_join)
time_res_annot <- time_res %>% 
  left_join(dplyr::select(annot,ID,true_name,true_biotype), by = c("geneid" = "ID"))
#write.table(time_res_annot, here::here(str_c(res_path, "tables/EDAseq_normalized results comparing by time WT infection.csv")),
 #          sep = ",", row.names = F)

res45vsn <- results(dds, contrast = c("group","WT45", "N45"))
res90vsn <- results(dds, contrast = c("group","WT90", "N90"))
res180vsn <- results(dds, contrast = c("group","WT180", "N180")) #Some significant differences shown
time_resvsn_annot <- list(res45vsn,res90vsn,res180vsn) %>% 
  map(as.data.frame) %>% 
  map2(c(".45",".90",".180"), ~rename_with(.x, .fn = function(colname) str_c(colname,.y))) %>% 
  map(rownames_to_column, var = "geneid") %>% 
  reduce(left_join) %>% 
  left_join(dplyr::select(annot, ID, true_name, true_biotype), by = c("geneid" = "ID"))
#write.table(time_resvsn_annot, here::here(str_c(res_path, "tables/EDAseq_normalized results comparing  WT and EscN infection.csv")),
 #          sep = ",", row.names = F)
```
Check overlap of DE genes between induced by time and by working T3SS
```{r de-overlap}
#Create a list of DE genes for each comparison
per_time_de <- list(res45,res90,res180, res45vsn, res90vsn, res180vsn) %>% 
  set_names(str_c("vs", rep(c("Planktonic","dEscN"), each = 3), rep(c(45,90,180), times = 2), sep = "_")) %>% 
  map(as.data.frame) %>% 
  map(~filter(.x, abs(log2FoldChange) >= 0.5, padj < 0.05)) %>% 
  map(rownames_to_column, var = "geneid") %>% 
  map(left_join, dplyr::select(annot, ID, true_name, true_biotype), by = c("geneid" = "ID")) %>% 
  map(pull, true_name)
#Create a list of DE genes that were significant at any point either against time or 
general_de <- list(reduce(per_time_de [c(1:3)], union), reduce(per_time_de [c(4:6)], union)) %>% 
  set_names(c("vs Time","vs dEscN"))
#Create and plot euler fits for these genes 
per_time_de_fit <- euler(per_time_de, shape = "ellipse")
general_de_fit <- euler(general_de, shape = "circle")

#pdf(here::here(str_c(res_path, "Euler of DE genes between comparisons.pdf")), width = 8, height = 6)
plot(per_time_de_fit, main = "Intersection of DE genes")
plot(general_de_fit, quantities = list(type = c("counts","percent")), main = "Union of DE genes for category")
#dev.off()
```


Analyze loadings of the principal components in the data
```{r pcatools, fig.width=8, fig.height=6.3}
#Annotate the gene id to more meaningful gene names
cnts <- assay(rld) 
cnts_annot <- cnts %>% 
  as_tibble(rownames = "ID") %>% 
  select(ID) %>% 
  left_join(dplyr::select(annot,ID,true_name))
#Confirm the order of rows is the same and can move onto symbol mapping
all(rownames(cnts) == cnts_annot$ID)
rownames(cnts) <- cnts_annot$true_name
p <- pca(cnts,colData(dds), removeVar = 0.1)
p_coord_meta <- p$rotated %>% 
  merge(as.data.frame(colData(rld)), by=0) 
#10 PCs explain 75% of variance, so just focus on those for linear regression
pc_to_choose <- 10
#Create the lists to store information on the linear model performance
fit_list <- set_names(vector(mode = "list", length = pc_to_choose),
                              nm = str_c("PC",1:pc_to_choose))
signi_list <- set_names(vector(mode = "list", length = pc_to_choose),
                              nm = str_c("PC",1:pc_to_choose))
#Generally, we are testing to see if metadata dictate PC coordinates
for (PC in seq_along(fit_list)) {
  fit_list [[PC]] <- set_names(vector(mode = "list", length = 3),
                               nm = c("group", "timepoint", "genotype"))
  signi_list [[PC]] <- set_names(vector(mode = "list", length = 3),
                                 nm = c("group", "timepoint", "genotype"))
  for (metavar in c("group", "timepoint", "genotype")) {
    frm <- as.formula(str_c("PC",PC, " ~ ", metavar))
    fit_list [[PC]] [[metavar]] <- lm(frm, data = p_coord_meta)
    sum_of_fit <- summary(fit_list [[PC]] [[metavar]])
    signi_list [[PC]] [[metavar]] <- sum_of_fit$coefficients %>% 
      as.data.frame() %>% 
      mutate(R_squared = sum_of_fit$r.squared,
             PC = str_c("PC",PC),
             metavar = metavar) %>% 
      rownames_to_column(var = "coef")
  }
}
#Summarise liner model performance and plot the results
signi_list_reduced <- signi_list %>% 
  map_depth(1,reduce,bind_rows) %>% 
  reduce(bind_rows) %>% 
  filter(coef != "(Intercept)") %>% 
  dplyr::rename(p = `Pr(>|t|)`) %>%  
  mutate(signi = if_else(p<0.05, "yes","no"),
         time = if_else(str_detect(coef,"group"), as.numeric(str_extract(coef, "[:digit:]+")), 0),
         coef = fct_reorder(coef, time),
         coef = fct_relevel(coef, "timepoint"))
#Plot the results of regression
ggplot(signi_list_reduced) +
  geom_tile(aes(y=coef,x=fct_relevel(PC,"PC10", after = Inf), fill = p)) +
  geom_text(data = filter(signi_list_reduced, signi == "yes"), aes(y=coef,x=fct_relevel(PC,"PC10", after = Inf)), 
            label = "*", color = "white", size = 7, nudge_y = -0.02) +
  labs(x = element_blank(), y = element_blank(), 
       title = "Linear regression analysis\nidentifying relationships between metadata and PC",
       caption = "For groups with genotype and time this shows\nsuccess in separation from WT EPEC at 0hr") +
  scale_y_discrete(labels = c(timepoint = "Time of Infection",
                              genotypeWT = "Effect of T3SS over time",
                              groupWT45 = "WT EPEC at 45 min",
                              groupWT90 = "WT EPEC at 90 min",
                              groupWT180 = "WT EPEC at 180 min",
                              groupN0 = bquote(Delta*"EscN EPEC before infection"),
                              groupN45 = bquote(Delta*"EscN EPEC at 45 min"),
                              groupN90 = bquote(Delta*"EscN EPEC at 90 min"),
                              groupN180 = bquote(Delta*"EscN EPEC at 180 min"))) +
  scale_fill_carto_c(palette = "BluYl", direction = -1, name = "p-value") +
  theme_minimal(base_size = 16) +
  theme(panel.border = element_rect(fill = NA, color = "black"),
        axis.ticks.x = element_line(color = "black"),
        axis.ticks.y = element_line(color = "black"),
        panel.grid.major = element_blank(),
        axis.text = element_text(color = "black"),
        axis.text.x = element_text(angle = 45, vjust = 0.6),
        title = element_text(size = 12),
        legend.title = element_text(size = 19))
#ggsave(here::here(str_c(res_path,"Linear regression of metadata and PCA with respect to RUV.png")),width = 8, height = 6.66, dpi = 1080)

signi_list_rsquare <- signi_list_reduced %>% 
  distinct(metavar, PC, .keep_all = T) %>% 
  mutate(metavar = str_to_title(metavar))

ggplot(signi_list_rsquare) +
  geom_tile(aes(y=metavar,x=fct_relevel(PC,"PC10", after = Inf), fill = R_squared)) +
  geom_text(data = filter(signi_list_rsquare, signi == "yes"), aes(y=metavar,x=fct_relevel(PC,"PC10", after = Inf)), 
            label = "*", color = "white", size = 7, nudge_y = -0.02) +
  labs(x = element_blank(), y = element_blank(), 
       title = "Linear regression analysis\nidentifying relationships between metadata and PC") +
  scale_y_discrete(labels = c(timepoint = "Time of Infection",
                              genotypeWT = "Effect of T3SS over time",
                              groupWT45 = "WT EPEC at 45 min",
                              groupWT90 = "WT EPEC at 90 min",
                              groupWT180 = "WT EPEC at 180 min",
                              groupN0 = bquote(Delta*"EscN EPEC before infection"),
                              groupN45 = bquote(Delta*"EscN EPEC at 45 min"),
                              groupN90 = bquote(Delta*"EscN EPEC at 90 min"),
                              groupN180 = bquote(Delta*"EscN EPEC at 180 min"))) +
  scale_fill_carto_c(palette = "BluYl", direction = 1, name = bquote("R"^2*"-value")) +
  theme_minimal(base_size = 16) +
  theme(panel.border = element_rect(fill = NA, color = "black"),
        axis.ticks.x = element_line(color = "black"),
        axis.ticks.y = element_line(color = "black"),
        panel.grid.major = element_blank(),
        axis.text = element_text(color = "black"),
        axis.text.x = element_text(angle = 45, vjust = 0.6),
        title = element_text(size = 12),
        legend.title = element_text(size = 19))
#ggsave(here::here(str_c(res_path,"Linear regression of metadata and PCA with respect to RUV R-square.png")),width = 8, height = 6.66, dpi = 1080)
```



```{r pcatools-old, fig.width=8, fig.height=6.3}
#pdf(here::here(str_c(res_path,"PCA tools analysis of PCA.pdf")), width = 8, height = 6.3)
screeplot(p)
biplot(p, colby = "timepoint", shape = "genotype",gridlines.major = FALSE)
biplot(p, x = "PC1", y = "PC3", colby = "timepoint", shape = "genotype",gridlines.major = FALSE)
plotloadings(p,  rangeRetain = 0.01,
    labSize = 4.0,col = carto_pal(3,diverg_pal_choice),
    caption = "Top 1% variables")
#dev.off()
```
Make volcano plots of differentially expressed genes
```{r volcano}
#shrink results
res45lfc <- lfcShrink(dds, coef="group_WT45_vs_WT0", type="apeglm", res = res45)
res90lfc <- lfcShrink(dds, coef="group_WT90_vs_WT0", type="apeglm", res = res90)
res180lfc <- lfcShrink(dds, coef="group_WT180_vs_WT0", type="apeglm", res = res180)
#Put into a list
results_list <- list(res45lfc, res90lfc,res180lfc)
#Prepare the results list for plotting
res_df_list <- results_list %>% 
  map(as.data.frame) %>% 
  map(rownames_to_column, var = "gene") %>% 
  map(mutate, signi = if_else(padj < 0.05 & abs(log2FoldChange) >= 0.5, "yes","no"))
#Create the function for making a nice volcano plot
volcano_ggplot <- function(res_df, title = "", to_label = F, n_lab = 10) {
  p_empty <- ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj), fill = signi, color = signi)) +
    geom_hline(yintercept = -log10(0.05), lty = 2, lwd = 0.8, color = "grey30") +
    geom_vline(xintercept = -0.5, lty = 2, lwd = 0.8, color = "grey30") +
    geom_vline(xintercept = 0.5, lty = 2, lwd = 0.8, color = "grey30") +
    geom_point(size = 4, shape = 21, alpha = 0.7) +
    scale_color_manual(values = c("darkgrey", carto_pal(7,diverg_pal_choice) [7])) +
    scale_fill_manual(values = c("darkgrey", carto_pal(7,diverg_pal_choice) [7])) +
    guides(color = 'none', fill = 'none') +
    labs(x = bquote(~Log[2] ~ "fold change"),
         y = bquote(~-Log[10] ~ italic(P)),
         title = title) +
    theme_prism(base_size = 24) +
    theme(title = element_text(size = 10),
          axis.title = element_text(size = 24))
  if(to_label) {
    res_df_top <- res_df %>% 
      filter(signi == "yes") %>% 
      arrange(desc(abs(log2FoldChange))) %>% 
      slice_head(n=n_lab)
    p_labelled <- p_empty + 
      geom_label_repel(data = res_df_top, aes(x = log2FoldChange, y = -log10(padj), label = gene),
                       size = 4, max.overlaps = Inf, color = "black", fill = "white")
    return(p_labelled)
  } else {
    return(p_empty)
  }
}
#Print the volcano plots - each should get its own png file
volcano_list <- map2(res_df_list, 
                     c("WT infection at 45min","WT infection at 90min", "WT infection at 180min"),
                     volcano_ggplot)
#walk2(volcano_list, c("WT infection at 45min","WT infection at 90min", "WT infection at 180min"),
 #     ~ggsave(here::here(str_c(res_path,"Volcano_plot of ",.y,".png")),plot = .x), width = 8, height = 6, dpi = 1080)

#pdf(here::here(str_c(res_path,"MA plots.pdf")), width = 8, height = 6.3)
plotMA(res45, main = "WT infection 45 min vs 0min \nResults without shrinking")
plotMA(res45lfc, main = "WT infection 45 min vs 0min \nResults after apeglm shrinking")
plotMA(res90, main = "WT infection 90 min vs 0min \nResults without shrinking")
plotMA(res90lfc, main = "WT infection 90 min vs 0min \nResults after apeglm shrinking")
plotMA(res180, main = "WT infection 180 min vs 0min \nResults without shrinking")
plotMA(res180lfc, main = "WT infection 180 min vs 0min \nResults after apeglm shrinking")
#dev.off()
```
Apply clusterprofiler to differentially expressed data
EPEC code in KEGG is ecg

For GSEA, use all genes, not just DE - https://hbctraining.github.io/Training-modules/DGE-functional-analysis/lessons/functional_analysis_2019.html

```{r clusterprofiler}
#library(pathview)
#If KEGG is giving troubles with clusterprofiler run this, can switch to wget instead of auto:
library(R.utils)
R.utils::setOption("clusterProfiler.download.method","wget")

#Prep a list of each gene lists
gl_all_df_list <- list(res45lfc,res90lfc,res180lfc, res45vsn,res90vsn, res180vsn) %>% 
  set_names(str_c("vs", rep(c("Time","dEscN"), each = 3), rep(c(45,90,180), times = 2), sep = "_")) %>% 
  map(as_tibble, rownames = "ID") 
gl_df_list <- gl_all_df_list %>% 
  map(~filter(.x,padj < 0.05, abs(log2FoldChange) > 0.5)) %>% 
  map(~arrange(.x,desc(log2FoldChange)))
#Get the genelists prepared
gl_gene_list <- gl_df_list %>% 
  map(~set_names(pull(.x,var = "log2FoldChange"), pull(.x, var = "ID")))
gse_gene_list <- gl_all_df_list %>% 
  map(~arrange(.x,desc(log2FoldChange))) %>% 
  map(~set_names(pull(.x,var = "log2FoldChange"), pull(.x, var = "ID")))
#Run the ora and gsea analyses
ora_list <- gl_gene_list %>% 
  map(~enrichKEGG(gene = names(.x),
                  organism     = 'ecg',
                  universe = rownames(res180lfc),
                  pvalueCutoff = 0.05,
                  qvalueCutoff = 0.05))
gse_list <- gse_gene_list %>% 
  map(~gseKEGG(geneList     = .x,
               organism     = 'ecg',
               pvalueCutoff = 0.05,
               verbose      = F))


#Also divide genes by up and down regulation
gl_updown_list <- gl_gene_list
for (comp in seq_along(gl_gene_list)) {
  genes_vector <- gl_gene_list [[comp]]
  gl_updown_list [[comp]] <- vector(mode = "list", length = 2) %>% 
    set_names(c("up","down"))
  gl_updown_list [[comp]] [[1]] <- genes_vector [genes_vector > 0]
  gl_updown_list [[comp]] [[2]] <- genes_vector [genes_vector < 0]
}
#Repeat ora and gse analyses
ora_updown_list <- gl_updown_list %>% 
  map_depth(2,~enrichKEGG(gene = names(.x),
                  organism     = 'ecg',
                  universe = rownames(res180lfc),
                  pvalueCutoff = 0.05,
                  qvalueCutoff = 0.05))
ora_updown_df <- ora_updown_list %>% 
  map_depth(2,as.data.frame) %>% 
  map(~map2(.x, c("Up","Down"), 
            ~mutate(.x, Direction = .y))) %>% 
  map(reduce, bind_rows) %>% 
  map(mutate, geneRatio = as.numeric(DOSE::parse_ratio(GeneRatio)))
```


```{r clusterprofiler-plot, eval=FALSE, fig.width=7, fig.height=5}
#Visualize all the results
#pdf(here::here(str_c(res_path,"Functional profiling plots.pdf")), width = 7, height = 5)
walk2(ora_list, names(ora_list),
      ~print(dotplot(.x) + scale_color_carto_c(palette = cont_pal_choice, direction = -1) +
               labs(title = str_c("whole ORA ",.y))))
walk2(keep(gse_list, ~nrow(.)>0), names(keep(gse_list, ~nrow(.)>0)),
      ~print(dotplot(.x, x = "NES") + scale_color_carto_c(palette = cont_pal_choice, direction = -1) +
               labs(title = str_c("GSEA ",.y))))

walk2(ora_updown_df, names(ora_updown_df),
      ~print(ggplot(.x, 
                    aes(x=geneRatio, y = fct_reorder(Description, geneRatio),
                        color = p.adjust, fill = p.adjust, size = Count, shape = Direction)) +
               geom_point(color = "black") +
               scale_shape_manual(values = c("Up" = 24, "Down" = 25)) +
               scale_size_continuous(range = c(4,8)) +
               scale_color_carto_c(palette = cont_pal_choice, direction = -1) +
               scale_fill_carto_c(palette = cont_pal_choice, direction = -1) +
               theme_minimal(base_size = 12) +
               guides(shape = guide_legend(override.aes = list(size = 4.5, fill = carto_pal(2,cont_pal_choice) [2]))) +
               labs(x="GeneRatio", y=element_blank(), title = .y) +
               theme(panel.border = element_rect(fill = NA, color = "black"),
                     axis.text = element_text(color = "black"))))

#dev.off()
#print out all the clusterprofiler results
#walk2(c(ora_list,keep(gse_list, ~nrow(.)>0), ora_updown_df), 
 #     c(str_c("whole_ORA",names(ora_list)), str_c("GSEA_", names(keep(gse_list, ~nrow(.)>0))), names(ora_updown_df)),
  #    ~write_csv(as.data.frame(.x), file = here::here(str_c(res_path,"tables/",.y,".csv"))))
```
Finally make a heatmap for top DE genes
```{r heatmap, fig.width=8, fig.height=6}
heat_mat <- assay(rld)
#Select 100 most DE genes from WT infection at 3hr
top_genes <- time_res_annot %>% 
  filter(padj.180 < 0.05, abs(log2FoldChange.180) >= 0.5) %>% 
  arrange(desc(abs(log2FoldChange.180))) %>% 
  slice_head(n=100)
heat_mat_select <- heat_mat [top_genes$geneid,]

rownames(heat_mat_select) <- top_genes$true_name
mat <- heat_mat_select - rowMeans(heat_mat_select)
#Set the breaks to be able to center the heatmap colors around 0
mycolors <- rcartocolor::carto_pal(name = diverg_pal_choice)
pal_length <- 255
mybreaks <-  c(seq(min(mat), 0, length.out=ceiling(pal_length/2) + 1), 
                    seq(max(mat)/pal_length, max(mat), length.out=floor(pal_length/2)))
color_scheme <- colorRampPalette(mycolors) (pal_length)


#For some reason fct_recode did not want to work the delta sign, so had to work around it
df <- as.data.frame(colData(rld)) %>%  
  mutate(`Infection status` = fct_recode(genotype, "T3SS-deficient EPEC" = "N",
                                            "Wild-type EPEC" = "WT"), .keep = 'unused') %>% 
  mutate(Timepoint = fct_recode(as.factor(timepoint), "Prior to infection" = "0", "45 min" = "45",
                                "90 min" = "90","180 min" = "180")) %>% 
  select(-group, -timepoint)

pheatmap(mat, annotation_col=df, 
         annotation_colors = list("Infection status" = c("Wild-type EPEC" =carto_pal(12,discrete_pal_choice) [2],
                                                         "T3SS-deficient EPEC" = carto_pal(12,discrete_pal_choice) [1]),
                                  "Timepoint" = c("Prior to infection" = chosen_cont_colors [1],
                                                  "45 min" = chosen_cont_colors [2],
                                                  "90 min" = chosen_cont_colors [3],
                                                  "180 min" = chosen_cont_colors [4])),
         fontsize = 11, treeheight_row = 0, treeheight_col = 10, breaks = mybreaks,
         show_rownames = F, show_colnames = F, border_color = NA, fontsize_row = 3.5,
         color = color_scheme)
```

```{r heatmap-save, eval=FALSE}
pheatmap(mat, annotation_col=df, 
         annotation_colors = list("Infection status" = c("Wild-type EPEC" =carto_pal(12,discrete_pal_choice) [2],
                                                         "T3SS-deficient EPEC" = carto_pal(12,discrete_pal_choice) [1]),
                                  "Timepoint" = c("Prior to infection" = chosen_cont_colors [1],
                                                  "45 min" = chosen_cont_colors [2],
                                                  "90 min" = chosen_cont_colors [3],
                                                  "180 min" = chosen_cont_colors [4])),
         fontsize = 11, treeheight_row = 0, treeheight_col = 10, breaks = mybreaks,
         show_rownames = F, show_colnames = F, border_color = NA, fontsize_row = 3.5,
         color = color_scheme,
         width = 8, height = 6, 
         filename = here::here(str_c(res_path,"top 100 DE genes WT180.png")))
```

Next, make a heatmap of virulence genes (LEE) and plotCounts of map and other mitochondrial genes
Take LEE annotation from [Antonio's paper](https://www.sciencedirect.com/science/article/pii/S1369527420300011?via=ihub#fig0005)

```{r lee-annot, fig.width=8, fig.height=6}
#Select only LEE and T2SS genes

top_genes <- time_res_annot %>% 
  mutate(vir_island = case_when(
    true_name %in% c("ler","escE","cesA", "cesB","escK","escL","escR","escS","escT","escU") ~ "LEE1",
    true_name %in% c("cesD","escC","sepD","escJ","escI","espZ") ~ "LEE2",
    true_name %in% c("cesL","escV","escN","escO","escP","escQ","espH") ~ "LEE3",
    true_name %in% c("tir","cesT","eaeA") ~ "LEE5",
    true_name %in% c("sepL","espA","espD","espB","cesD2","escF","escG","espF") ~ "LEE4",
    str_detect(true_name,"gsp[:Alpha:]+") ~ "T2SS")) %>% 
  filter(!is.na(vir_island)) %>% 
  group_by(vir_island) %>% 
  arrange(desc(abs(log2FoldChange.180)), .by_group = T) 
heat_mat_lee <- heat_mat [top_genes$geneid,]

rownames(heat_mat_lee) <- top_genes$true_name
mat_lee <- heat_mat_lee - rowMeans(heat_mat_lee)
#Set the breaks to be able to center the heatmap colors around 0
mycolors <- rcartocolor::carto_pal(name = diverg_pal_choice)
pal_length <- 255
mybreaks <-  c(seq(min(mat), 0, length.out=ceiling(pal_length/2) + 1), 
                    seq(max(mat)/pal_length, max(mat), length.out=floor(pal_length/2)))
color_scheme <- colorRampPalette(mycolors) (pal_length)


pheatmap(mat_lee, annotation_col=df, 
         annotation_colors = list("Infection status" = c("Wild-type EPEC" =carto_pal(12,discrete_pal_choice) [2],
                                                         "T3SS-deficient EPEC" = carto_pal(12,discrete_pal_choice) [1]),
                                  "Timepoint" = c("Prior to infection" = carto_pal(4,cont_pal_choice) [1],
                                                  "45 min" = carto_pal(4,cont_pal_choice) [2],
                                                  "90 min" = carto_pal(4,cont_pal_choice) [3],
                                                  "180 min" = carto_pal(4,cont_pal_choice) [4])),
         fontsize = 10, treeheight_row = 0, treeheight_col = 10, breaks = mybreaks, cluster_rows = F,
         show_rownames = T, show_colnames = F, border_color = NA, fontsize_row = 8,
         color = color_scheme)
```
```{r lee-heatmap-save, eval=FALSE}
pheatmap(mat_lee, annotation_col=df, 
         annotation_colors = list("Infection status" = c("Wild-type EPEC" =carto_pal(12,discrete_pal_choice) [2],
                                                         "T3SS-deficient EPEC" = carto_pal(12,discrete_pal_choice) [1]),
                                  "Timepoint" = c("Prior to infection" = carto_pal(4,cont_pal_choice) [1],
                                                  "45 min" = carto_pal(4,cont_pal_choice) [2],
                                                  "90 min" = carto_pal(4,cont_pal_choice) [3],
                                                  "180 min" = carto_pal(4,cont_pal_choice) [4])),
         fontsize = 10, treeheight_row = 0, treeheight_col = 10, breaks = mybreaks, cluster_rows = F,
         show_rownames = T, show_colnames = F, border_color = NA, fontsize_row = 8,
         color = color_scheme,
          width = 8, height = 6, 
         filename = here::here(str_c(res_path,"heatmap virulence genes.png")))
```

Finally, plot counts for mitochondrial-related effectors over time

```{r mito-counts, fig.width=7, fig.height=5}
mito_ids <- time_res_annot %>% 
  filter(true_name %in% c("E2348C_3942","espH","espF","espZ", "E2348C_0718", "E2348C_3231",
                          "E2348C_1444", "E2348C_1445", "E2348C_1044")) %>% 
  mutate(true_name = str_replace(true_name,"E2348C_3942","map"),
         true_name = fct_recode(true_name, "nleH1" = "E2348C_0718", "nleH2" = "E2348C_1444",
                                "nleB" = "E2348C_3231","nleF" = "E2348C_1445", "nleD" = "E2348C_1044")
         )
mito_counts <- set_names(mito_ids$geneid, mito_ids$true_name) %>% 
  map(~plotCounts(dds, gene = .x, returnData = T, intgroup = "timepoint")) %>% 
  map2(mito_ids$true_name, ~mutate(.x, true_name = .y)) %>% 
  map(rownames_to_column,var = "sample") %>% 
  reduce(bind_rows) %>% 
  left_join(as_tibble(pData, rownames = "sample")) %>% 
  mutate(genotype = fct_recode(genotype, "Wild-Type EPEC" = "WT", "T3SS Mutant" = "N"),
         lee = if_else(str_detect(true_name, "nle"), "Non-LEE", "LEE")
         )

p_lee <- ggplot(filter(mito_counts, lee == "LEE"), aes(x=timepoint,y=count, color = true_name)) +
  geom_line(stat = "summary", fun = "mean", linewidth = 1) +
  stat_summary(geom = "pointrange", fun.data = "mean_se",size = 0.7) + 
  scale_x_continuous(breaks = c(0,45,90,180)) +
  scale_color_carto_d(palette = discrete_pal_choice, name = "Mitochondrial\neffector") +
  labs(x = "Timepoint (min)", y = "Normalized Counts") +
  facet_wrap(~genotype, ncol = 2, scales = "free_y")
p_nolee <- ggplot(filter(mito_counts, lee != "LEE"), aes(x=timepoint,y=count, color = true_name)) +
  geom_line(stat = "summary", fun = "mean", linewidth = 1) +
  stat_summary(geom = "pointrange", fun.data = "mean_se",size = 0.7) + 
  scale_x_continuous(breaks = c(0,45,90,180)) +
  scale_color_carto_d(palette = "Safe", name = "Apoptotic\neffector") +
  labs(x = "Timepoint (min)", y = "Normalized Counts") +
  facet_wrap(~genotype, ncol = 2, scales = "free_y")
```


```{r mito-counts-plot, fig.width=7, fig.height=10}
p_lee/p_nolee
#ggsave(here::here(str_c(res_path,"Mitochondrial,apoptotic effector counts over time.png")), width = 7, height = 10, dpi = 1080)
```

Finally export normalized counts for correlations
```{r export-counts, eval=FALSE}
norm_cnts <- counts(dds, normalized = T) %>% 
  as_tibble(rownames = "geneid")
#write_csv(norm_cnts,here::here(str_c(res_path,"tables/normalized_counts.csv")))
```







