---
title: "Final clean analysis of short-term infection RNA-seq"
output: 
  html_notebook: 
    fig_width: 6
    fig_height: 5
editor_options: 
  chunk_output_type: inline
---

```{r results='hide', message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(include = TRUE)
knitr::opts_chunk$set(fig.width = 6)
knitr::opts_chunk$set(fig.height = 5)
library(greekLetters)
library(RUVSeq)
library(EDASeq)
library(DESeq2)
library(PCAtools)
library('org.Hs.eg.db')
library(EGSEA)
library(limma)
library(pheatmap)
library(eulerr)
library(UpSetR)
library(clusterProfiler)
library(tidyverse)
library(ggprism)
library(rcartocolor)
```

Define all helper functions and color choices
```{r helper, include=TRUE}
theme_set(theme_prism(base_size = 14))
#Set color choices from CARTOcolors
discrete_pal_choice <- "Bold"
cont_pal_choice <- "BluYl"
diverg_pal_choice <- "Geyser"
#Set color vectors
cont_colors <- carto_pal(name = cont_pal_choice)
chosen_cont_colors <- cont_colors [c(2,4,6,7)]
#Helper functions
calc_relative <- function(x) x/sum(x) * 100
#Function to assign rownames to a matrix
assign_rownames <- function(mat, row_names) {
  rownames(mat) <- row_names
  return(mat)
}
#Set up paths for data, and results
data_path <- "short_infection_human_reads/data/"
res_path <- "short_infection_human_reads/results/final clean figures/"
```

#Read in the count data and rename the columns to fit the sample names

```{r readorigstar}
counts <- read.table(here::here(str_c(data_path,"STAR_reads_updated.txt")),header = F, sep = "\t", stringsAsFactors = F)
#Next manipulate the table to remove redundant columns and only keep relevant columns (1 columns with gene names, and the rest with counts for reverse stranded library
counts_simple <- counts [,seq(from=4,to=96,by=4)]
colnames(counts_simple) <- c("4UIN", "4UIWT", "490WT", "490N", "4180N", "4180WT", "5UIN", "5UIWT", "545WT", "545N", "590WT", "590N", "5180N", "5180WT", "6UIN", "6UIWT", "645WT", "645N", "690WT", "690N", "6180N", "6180WT", "445N", "445WT")
#Prior analysis revealed that 6UIN is an outlier by PCA plot
counts_simple <- counts_simple %>% 
  dplyr::select(-`6UIN`)
counts_simple$Gene_ID <- counts$V1

filter <- apply(dplyr::select(counts_simple,-Gene_ID),1,function(x) mean(x)>5)
counts_simple <- counts_simple [filter,]

#write.table(counts_simple,file=here::here(str_c(res_path,"tables/only_reverse_strand_counts_cleaned.txt")),sep = "\t", row.names = F)
```

```{r vis-align-stats}
#Visualize alignment statistics
counts_stats <- counts_simple %>% 
  dplyr::slice(1:4) %>% 
  pivot_longer(-Gene_ID, values_to = "numbers", names_to = "mapping")
mapped <-  counts_simple %>%
  dplyr::select(-Gene_ID) %>%
  summarise(across(everything(), sum)) %>% 
  pivot_longer(everything(),values_to = "numbers", names_to = "mapping") %>% 
  mutate(Gene_ID = "Mapped") %>% 
  bind_rows(counts_stats) %>% 
  mutate(Gene_ID = fct_relevel(str_to_title(str_replace(Gene_ID, "N_","")), "Mapped"))

#Raw counts stats
raw_count_plot <- ggplot(mapped, aes(x=mapping, y=numbers, fill = Gene_ID)) +
  geom_bar(stat = "identity") + scale_fill_carto_d(palette = discrete_pal_choice) +
  labs(x = "", y = "Raw counts") +theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6)) + 
  guides(fill = guide_legend(title = "Mapping Status"))
#Relative count stats
rel_map <- mapped %>% 
  pivot_wider(Gene_ID, names_from = mapping, values_from = numbers) %>%  
  mutate(across(-Gene_ID, calc_relative)) %>% 
  pivot_longer(-Gene_ID,values_to = "numbers", names_to = "mapping")
rel_count_plot <-ggplot(rel_map, aes(x=mapping, y=numbers, fill = Gene_ID)) +
  geom_bar(stat = "identity") + scale_fill_carto_d(palette = discrete_pal_choice) +
  labs(x = "", y = "Proportion of counts") +
  theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6)) + 
  guides(fill = guide_legend(title = "Mapping Status"))
#pdf(file = here::here(str_c(res_path,"mapping stats.pdf")), width = 8, height = 6)
raw_count_plot
rel_count_plot
#dev.off()
```

Calculate statistics of types of transcripts mapped

```{r map-by-type}
geneids <- mapIds(org.Hs.eg.db, keys = str_replace_all(counts_simple$Gene_ID [-1:-4], "\\.[1-9]([0-9])?", ""), 
                  keytype = "ENSEMBL", column = "GENETYPE", multiVals = "first") %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "Gene_ID")
c_type <- counts_simple %>% 
  dplyr::slice(-1:-4) %>% 
  mutate(Gene_ID = str_replace_all(Gene_ID, "\\.[1-9]([0-9])?", "")) %>% 
  left_join(geneids) %>% 
  dplyr::select(-Gene_ID) %>%
  rename_with(function(x) "Type",last_col()) %>% 
  group_by(Type) %>% 
  summarise(across(everything(),sum)) %>% 
  mutate(across(-Type, calc_relative)) %>% 
  pivot_longer(-Type, values_to = "numbers", names_to = "mapping") %>% 
  mutate(Type = str_replace_na(Type, replacement = "other"))
rel_type_plot <- ggplot(c_type, aes(x=mapping, y=numbers, fill = fct_reorder(Type,numbers, .desc = T))) +
  geom_bar(stat = "identity") + scale_fill_carto_d(palette = discrete_pal_choice) +
  labs(x = "", y = "Proportion of genes") +
  theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6)) + 
  guides(fill = guide_legend(title = "Gene type"))
#pdf(file = here::here(str_c(res_path,"mapping stats by gene type.pdf")), width = 8, height = 6)
rel_type_plot
#dev.off()
```


Apply EDAseq normalization to the data

```{r edaseq}
#Use EDAseq to normalize the libraries
c_mat <- counts_simple %>% 
                     dplyr::select(-dplyr::last_col()) %>% 
                     dplyr::slice(-1:-4) %>% 
                     as.matrix()
rownames(c_mat) <- counts_simple$Gene_ID [-1:-4]
pData <- data.frame(genotype=vector(mode = "logical", length = ncol(c_mat)),
                    timepoint = vector(mode = "logical", length = ncol(c_mat)),
                    samples = colnames(c_mat),
                    row.names = colnames(c_mat))
pData <- pData %>% 
  mutate(genotype = case_when(
    str_detect(samples, "WT") ~ "WT",
    str_detect(samples, "UIN") ~ "WT",
    str_detect(samples, "N") ~ "N"
  ), timepoint = case_when(
    str_detect(samples, "UI") ~ 0,
    str_detect(samples, "45") ~ 45,
    str_detect(samples, "90") ~ 90,
    str_detect(samples, "180") ~ 180,
  )) %>% dplyr::select(-samples)

data <- newSeqExpressionSet(counts=c_mat,
                            phenoData=pData)
#Normalize the data next
dataNorm <- betweenLaneNormalization(data, which="upper") 
#Plot the counts
boxplot(data, main = "Unnormalized counts", col = "dodgerblue3")
boxplot(dataNorm, main = "Normalized counts", col = "dodgerblue3")
#For DE we will create this analysis with offset values
dataOffset <- betweenLaneNormalization(data,
                                       which="upper",offset=TRUE)
```
Next step is to use normalized counts in the DESEq2 analysis of differential expression
```{r deseq, fig.width=7, fig.height=5}
#Prepare the DESEq2 object and apply EDASeq normalization factors
pData$group <- factor(paste0(pData$genotype,pData$timepoint))
pData <- pData %>% 
  mutate(genotype = case_when(
    timepoint == 0 ~ "Uninfected",
    TRUE ~ genotype
  ))
dds <- DESeqDataSetFromMatrix(countData = counts(dataOffset),
                              colData = pData,
                              design = ~ group)
dds$group <- relevel(dds$group, ref= "WT0")
normFactors <- exp(-1 * offst(dataOffset))
normFactors <- normFactors / exp(rowMeans(log(normFactors)))
normalizationFactors(dds) <- normFactors

#Perform the DESEq2 analysis
dds <- DESeq(dds)
rld <- rlog(dds)

normal_PCA <- plotPCA(rld, returnData = T, intgroup = "group") %>% 
  mutate(timepoint = as.numeric(str_remove(group, "[:alpha:]+")),
         genotype = ifelse(timepoint == 0, "Uninfected", str_remove(group, "[:digit:]+"))) %>% 
  mutate(genotype = fct_relevel(genotype, "Uninfected", "WT", "N"))
  
#Plot the PCA graphs of the normalized counts
pca_plot <- ggplot(data = normal_PCA, aes(shape = genotype)) +
  scale_color_manual(values = chosen_cont_colors,breaks = c(0,45,90,180)) +
  #scale_color_carto_c(palette = cont_pal_choice, breaks = c(0,45,90,180)) +
  geom_point(mapping = aes(x = PC1, y = PC2, color = factor(timepoint)), size = 5) +
  labs(title = "EDAseq normalized DESeq2 PCA plot \nnot including the outlier",
       x = "PC1: 39% Variance",
       y = "PC2: 21% Variance") +
  guides(color = guide_legend(title = "Time of Infection (min)"),
         shape = guide_legend(title = "Infection Status")) +
  scale_shape_manual(labels = c("WT" = "WT",
                                "N" = expression(Delta*italic("escN"))),
              values = c("Uninfected" = 15, "WT" = 16, "N" = 17)) +
  theme(legend.title = element_text(),
        legend.text.align = 0)
pca_plot
#ggsave(here::here(str_c(res_path,"PCA plot of normalized counts.png")), pca_plot, width = 7, height = 5, dpi = 1080)
#ggsave(here::here(str_c(res_path,"PCA plot of normalized counts.svg")), pca_plot, width = 7, height = 5)
```
Identify differentially expressed genes
```{r deseq-res}
res45 <- results(dds, contrast = c("group","WT45", "WT0"))
res90 <- results(dds, contrast = c("group","WT90", "WT0"))
res180 <- results(dds, contrast = c("group","WT180", "WT0"))
time_res <- list(res45,res90,res180) %>% 
  map(as.data.frame) %>% 
  map2(c(".45",".90",".180"), ~rename_with(.x, .fn = function(colname) str_c(colname,.y))) %>% 
  map(rownames_to_column, var = "geneid") %>% 
  reduce(left_join)
#write.csv(time_res, here::here(str_c(res_path,"tables/EDAseq_normalized results comparing by time WT infection.csv")), row.names = F)
```


```{r deseq-res-wt-n}
res90vsn <- results(dds, contrast = c("group","WT90", "N90"))
res180vsn <- results(dds, contrast = c("group","WT180", "N180")) #none showed significant differences
summary(res90vsn)
summary(res180vsn)
```
Analyze loadings of the principal components in the data
```{r pcatools, fig.width=9, fig.height=6}
#Create a mapping for ID to symbol
geneids <- mapIds(org.Hs.eg.db, keys = str_replace_all(rownames(counts(dds)), "\\.[1-9]([0-9])?", ""), 
                  keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")

cnts <- assay(rld) 
#Confirm that that the mapping is correct
all(str_replace_all(rownames(counts(dds)), "\\.[1-9]([0-9])?", "") == names(geneids))
#Rename rows that had a symbol name
newnames <- ifelse(is.na(geneids) | duplicated(geneids),
    names(geneids), geneids)
rownames(cnts) <- newnames
p <- pca(cnts,colData(dds), removeVar = 0.1)
#pdf(here::here(str_c(res_path,"PCA tools loadings of PCA.pdf")), width = 9, height = 6)
plotloadings(p,  rangeRetain = 0.01,
    labSize = 2.5, col = carto_pal(3,diverg_pal_choice) ,
    caption = "Top 1% variables") + theme(legend.title = element_text())
#dev.off()
```
Print out normalized counts
```{r norm-counts-print, eval=FALSE}
norm_cnts <- counts(dds, normalized = T) %>% 
  as_tibble(rownames = "geneid") %>% 
  mutate(ensemble_geneid = str_replace_all(geneid, "\\.[1-9]([0-9])?", "")) %>% 
  left_join(enframe(geneids), by = c("ensemble_geneid" = "name")) %>% 
  mutate(true_name = if_else(is.na(value)|duplicated(value), ensemble_geneid, value))
#write_csv(norm_cnts,here::here(str_c(res_path, "tables/normalized counts.csv")))
```


Make volcano plots of differentially expressed genes
```{r volcano}
#shrink results
res45lfc <- lfcShrink(dds, coef="group_WT45_vs_WT0", type="apeglm", res = res45)
res90lfc <- lfcShrink(dds, coef="group_WT90_vs_WT0", type="apeglm", res = res90)
res180lfc <- lfcShrink(dds, coef="group_WT180_vs_WT0", type="apeglm", res = res180)

#Create a mapping for ID to symbol
geneids <- mapIds(org.Hs.eg.db, keys = str_replace_all(rownames(res180lfc), "\\.[1-9]([0-9])?", ""), 
                  keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first") %>% 
  as.data.frame() %>% 
  bind_cols(rownames(res180lfc)) %>% 
  dplyr::select("symbol" = 1, "rowname" = `...2`)
#Get nice names or keep ensembl IDs
naOrDup <- is.na(geneids$symbol) | duplicated(geneids$symbol)
geneids$gene_name <- ifelse(naOrDup, rownames(geneids), geneids$symbol)
results_list <- list(res45lfc, res90lfc,res180lfc)
#Replace the rownames with gene names
for (i in seq_along(results_list)) {rownames(results_list [[i]]) <- geneids$gene_name}
#Prepare the results list for plotting
res_df_list <- results_list %>% 
  map(as.data.frame) %>% 
  map(rownames_to_column, var = "gene") %>% 
  map(mutate, signi = if_else(padj < 0.05 & abs(log2FoldChange) >= 0.5, "yes","no"))

#Create the function for making a nice volcano plot
volcano_ggplot <- function(res_df, title = "", to_label = F, n_lab = 10) {
  p_empty <- ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj), fill = signi, color = signi)) +
    geom_hline(yintercept = -log10(0.05), lty = 2, lwd = 0.8, color = "grey30") +
    geom_vline(xintercept = -0.5, lty = 2, lwd = 0.8, color = "grey30") +
    geom_vline(xintercept = 0.5, lty = 2, lwd = 0.8, color = "grey30") +
    geom_point(size = 4, shape = 21, alpha = 0.7) +
    scale_color_manual(values = c("darkgrey", carto_pal(7,diverg_pal_choice) [7])) +
    scale_fill_manual(values = c("darkgrey", carto_pal(7,diverg_pal_choice) [7])) +
    guides(color = 'none', fill = 'none') +
    labs(x = bquote(~Log[2] ~ "fold change"),
         y = bquote(~-Log[10] ~ italic(P)),
         title = title) +
    theme_prism(base_size = 24) +
    theme(title = element_text(size = 10),
          axis.title = element_text(size = 24))
  if(to_label) {
    res_df_top <- res_df %>% 
      filter(signi == "yes") %>% 
      arrange(desc(abs(log2FoldChange))) %>% 
      slice_head(n=n_lab)
    p_labelled <- p_empty + 
      geom_label_repel(data = res_df_top, aes(x = log2FoldChange, y = -log10(padj), label = gene),
                       size = 4, max.overlaps = Inf, color = "black", fill = "white")
    return(p_labelled)
  } else {
    return(p_empty)
  }
}
#Print the volcano plots - each should get its own png file
volcano_list <- map2(res_df_list, 
                     c("WT infection at 45min","WT infection at 90min", "WT infection at 180min"),
                     volcano_ggplot)
#Repeat the same with adding top 10 DE genes
volcano_list_labelled <- map2(res_df_list, 
                     c("WT infection at 45min","WT infection at 90min", "WT infection at 180min"),
                     volcano_ggplot, to_label = T, n_lab = 10)
#pdf(here::here(str_c(res_path,"Volcano plots.pdf")), width = 8, height = 6)
walk(volcano_list,print)
walk(volcano_list_labelled,print)
#dev.off()


#pdf(here::here(str_c(res_path,"MA plots.pdf")), width = 8, height = 6)
plotMA(res90, main = "WT infection 90 min vs 0min \nResults without shrinking")
plotMA(res90lfc, main = "WT infection 90 min vs 0min \nResults after apeglm shrinking")
plotMA(res180, main = "WT infection 180 min vs 0min \nResults without shrinking")
plotMA(res180lfc, main = "WT infection 180 min vs 0min \nResults after apeglm shrinking")
#dev.off()
```


Create a heatmap of top differentially expressed genes
```{r heatmap, fig.width=8, fig.height=6}
heat_mat <- assay(rld)
#Select 100 most DE genes from WT infection at 3hr
top_genes <- time_res %>% 
  filter(padj.180 < 0.05, abs(log2FoldChange.180) >= 0.5) %>% 
  arrange(desc(abs(log2FoldChange.180))) %>% 
  slice_head(n=100)
heat_mat_select <- heat_mat [top_genes$geneid,]
#Create a mapping for ID to symbol
geneids <- mapIds(org.Hs.eg.db, keys = str_replace_all(rownames(heat_mat_select), "\\.[1-9]([0-9])?", ""), 
                  keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first") %>% 
  as.data.frame() %>% 
  bind_cols(rownames(heat_mat_select)) %>% 
  dplyr::select("symbol" = 1, "rowname" = `...2`)
#Get nice names or keep ensembl IDs
naOrDup <- is.na(geneids$symbol) | duplicated(geneids$symbol)
geneids$gene_name <- ifelse(naOrDup, rownames(geneids), geneids$symbol)
rownames(heat_mat_select) <- geneids$gene_name
mat <- heat_mat_select - rowMeans(heat_mat_select)
#Set the breaks to be able to center the heatmap colors around 0
mycolors <- rcartocolor::carto_pal(name = diverg_pal_choice)
pal_length <- 255
mybreaks <-  c(seq(min(mat), 0, length.out=ceiling(pal_length/2) + 1), 
                    seq(max(mat)/pal_length, max(mat), length.out=floor(pal_length/2)))
color_scheme <- colorRampPalette(mycolors) (pal_length)


#For some reason fct_recode did not want to work the delta sign, so had to work around it
df <- as.data.frame(colData(rld)) %>%  
  mutate(`Infection Status` = fct_recode(genotype, "Mutant" = "N",
                                            "WT" = "WT","Uninfected" = "UI"), .keep = 'unused') %>% 
  mutate(`Time of Infection (min)` = fct_recode(as.factor(timepoint), "0" = "0", "45" = "45",
                                "90" = "90","180" = "180")) %>% 
  select(-group, -timepoint)
#Reorder the samples in direction of time and condition
#Create the ordering in the order of the operons and direction of time
sample_order_wt <- as_tibble(colData(rld), rownames = "sample") %>% 
  filter(genotype == "WT") %>% 
  arrange(timepoint, sample) %>% 
  pull(sample)
sample_order_n <- as_tibble(colData(rld), rownames = "sample") %>% 
  filter(genotype != "WT") %>% 
  arrange(desc(timepoint), sample) %>% 
  pull(sample)
#Plot the heatmap
pheatmap(mat [,c(sample_order_n, sample_order_wt)], annotation_col=df, 
         annotation_colors = list("Infection Status" = c("Uninfected" = carto_pal(12,discrete_pal_choice) [10],
                                                         "WT" =carto_pal(12,discrete_pal_choice) [2],
                                                         "Mutant" = carto_pal(12,discrete_pal_choice) [1]),
                                  "Time of Infection (min)" = c("0" = chosen_cont_colors [1],
                                                                "45" = chosen_cont_colors [2],
                                                                "90" = chosen_cont_colors [3],
                                                                "180" = chosen_cont_colors [4])),
         fontsize = 11, treeheight_row = 0, treeheight_col = 0, breaks = mybreaks,
         show_rownames = F, show_colnames = F, border_color = NA, cluster_cols = F, gaps_col = c(9,14),
         color = color_scheme)
```

```{r pheatmap-save, eval=FALSE}
pheatmap(mat [,c(sample_order_n, sample_order_wt)], annotation_col=df, 
         annotation_colors = list("Infection Status" = c("Uninfected" = carto_pal(12,discrete_pal_choice) [10],
                                                         "WT" =carto_pal(12,discrete_pal_choice) [2],
                                                         "Mutant" = carto_pal(12,discrete_pal_choice) [1]),
                                  "Time of Infection (min)" = c("0" = chosen_cont_colors [1],
                                                                "45" = chosen_cont_colors [2],
                                                                "90" = chosen_cont_colors [3],
                                                                "180" = chosen_cont_colors [4])),
         fontsize = 11, treeheight_row = 0, treeheight_col = 0, breaks = mybreaks,
         show_rownames = F, show_colnames = F, border_color = NA, cluster_cols = F, gaps_col = c(9,14),
         color = color_scheme, width = 8, height = 6, 
         filename = here::here(str_c(res_path,"top 100 DE genes WT180.png")))
```

```{r heatmap-function-top-genes, fig.width=8, fig.height=6}
#Read in annotation for the top variable genes
h_symbol <- read.gmt(here::here(str_c(data_path, "h.all.v7.5.1.symbols.gmt")))
#Do a quick clusterprofiler analysis to identify which pathways are overrepresented
#prep the universe of detected genes
universe_symbol <- mapIds(org.Hs.eg.db, keys = str_replace(rownames(res180), "\\.[1-9]([0-9])?", ""), keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")
#This is not statistically binding - more for viz sake
heatmap_profile <- enricher(gene = rownames(mat), pvalueCutoff = 0.05, qvalueCutoff = 0.1,
                            universe = universe_symbol, TERM2GENE = h_symbol)
#Get the only relevant results out
hm_func_df <- heatmap_profile@result %>% 
  filter(qvalue < 0.1)
#Print out the table for future references
#write_csv(hm_func_df,here::here(str_c(res_path, "tables/profiling of top 100 variable genes short infection.csv")))
#Create the annotation for the heatmap rows of interest
annot_row <- h_symbol %>% 
  filter(term %in% hm_func_df$ID) %>% 
  right_join(tibble(gene = rownames(mat)), by = "gene") %>% 
  filter(!is.na(term)) %>% 
  filter(str_detect(term, "HYPOXIA", negate = T)) %>% 
  mutate(gene = if_else(duplicated(gene), str_c(gene, " "),gene)) %>% 
  mutate(gene = if_else(duplicated(gene), str_c(gene, " "),gene)) %>% 
  column_to_rownames(var = "gene") %>% 
  mutate(term = str_remove(term, "HALLMARK_"),
         term = str_replace_all(term,"_", " ")) %>% 
  rename(`Gene Set` = term) 
#Now identify genes which have belong to several functions
ambi_genes <- str_subset(rownames(annot_row), "[:alnum:] $")

#Create a matrix for ambi_genes 
ambi_mat <- mat [str_remove(ambi_genes, " "),] %>%  
  assign_rownames(., ambi_genes)
#Now create the new matrix for all results
func_mat <- rbind(mat,ambi_mat) 
func_mat <- func_mat [rownames(func_mat) %in% rownames(annot_row),c(sample_order_n, sample_order_wt)]
#Now we can try and cluster within each category
cat_mats <- unique(annot_row$`Gene Set`) %>% 
  set_names(.,.) %>% 
  map(~filter(annot_row, `Gene Set` == .x)) %>% 
  map(~func_mat [rownames(.x),])
#Run distances and clustering for each matrix
cat_mats_clust <- cat_mats %>% 
  map(dist, method = "euclidean") %>% 
  map(hclust)
map(cat_mats_clust, plot)
#From these clusters extract labels for easy ordering of the heatmap
clustered_labs <- cat_mats_clust %>% 
  map(~.x$order) %>%   
  map2(cat_mats,~rownames(.y) [.x]) %>% 
  reduce(c)
 #Draw the heatmap
pheatmap(func_mat [clustered_labs,], annotation_col=df, annotation_row = annot_row,
         annotation_colors = list("Infection Status" = c("Uninfected" = carto_pal(12,discrete_pal_choice) [10],
                                                         "WT" =carto_pal(12,discrete_pal_choice) [2],
                                                         "Mutant" = carto_pal(12,discrete_pal_choice) [1]),
                                  "Time of Infection (min)" = c("0" = chosen_cont_colors [1],
                                                                "45" = chosen_cont_colors [2],
                                                                "90" = chosen_cont_colors [3],
                                                                "180" = chosen_cont_colors [4]),
                                  "Gene Set" = c("TNFA SIGNALING VIA NFKB" = carto_pal(12,"Safe") [1],
                                                 "P53 PATHWAY" = carto_pal(12,"Safe") [5],
                                                 "IL2 STAT5 SIGNALING" = carto_pal(12,"Safe") [2])),
         fontsize_row = 10, fontsize = 11, treeheight_row = 0, treeheight_col = 0, breaks = mybreaks, cluster_rows = F,
         show_rownames = T, show_colnames = F, border_color = NA, cluster_cols = F, gaps_col = c(9,14), 
         gaps_row = c(27, 34), legend = F, annotation_names_row = F, annotation_names_col = F,
         color = color_scheme)
```


```{r func-pheatmap-save, eval=FALSE}
pheatmap(func_mat [clustered_labs,], annotation_col=df, annotation_row = annot_row,
         annotation_colors = list("Infection Status" = c("Uninfected" = carto_pal(12,discrete_pal_choice) [10],
                                                         "WT" =carto_pal(12,discrete_pal_choice) [2],
                                                         "Mutant" = carto_pal(12,discrete_pal_choice) [1]),
                                  "Time of Infection (min)" = c("0" = chosen_cont_colors [1],
                                                                "45" = chosen_cont_colors [2],
                                                                "90" = chosen_cont_colors [3],
                                                                "180" = chosen_cont_colors [4]),
                                  "Gene Set" = c("TNFA SIGNALING VIA NFKB" = carto_pal(12,"Safe") [1],
                                                 "P53 PATHWAY" = carto_pal(12,"Safe") [5],
                                                 "IL2 STAT5 SIGNALING" = carto_pal(12,"Safe") [2])),
         fontsize_row = 10, fontsize = 11, treeheight_row = 0, treeheight_col = 0, breaks = mybreaks, cluster_rows = F,
         show_rownames = T, show_colnames = F, border_color = NA, cluster_cols = F, gaps_col = c(9,14), 
         gaps_row = c(27, 34), legend = F, annotation_names_row = F, annotation_names_col = F,
         color = color_scheme, width = 8, height = 6, 
         filename = here::here(str_c(res_path,"top 100 variable genes functional split.png")))
```



Apply EGSEA to differentially expressed data

```{r eggsea-prep}
#remove intercept since this is a factor. Intercept or no intercept does not make a difference
#That is what the 0+ means in the formula
design_matrix <- model.matrix(~0+pData$group) 
colnames(design_matrix) <- colnames(design_matrix) %>% str_replace("pData\\$group","") 
names_for_contr <- resultsNames(dds) [-1] %>%
  str_replace_all("group_","") %>% 
  str_replace_all("_vs_","-")
contrast_obj <- makeContrasts(contrasts = names_for_contr, levels = colnames(design_matrix))
group_obj <- as.factor(pData$group)
cnt_obj <- dataNorm@assayData$normalizedCounts
#finally prepare the geneID mapping tool
geneids <- mapIds(org.Hs.eg.db, keys = str_replace(rownames(cnt_obj), "\\.[1-9]([0-9])?", ""), keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first")

#Replace the rownames of the count matrix with ENTREZ ID
rownames(cnt_obj) <- unname(geneids)
#Remove NAs from the matrix (not every gene had an NCBI match)
cnt_obj_filt <- cnt_obj [!is.na(rownames(cnt_obj)),] 
#Deal with the remaining duplicates
dupes <-  c(which(duplicated(rownames(cnt_obj_filt)) | duplicated(rownames(cnt_obj_filt),fromLast = T)))
means <- data.frame(matrix(nrow = length(dupes))) %>% mutate(index = dupes)
#Select the duplicate with higher gene count
for (i in 1:length(dupes)) {
  means$avg [i] <- mean(cnt_obj_filt[dupes[i],])
  means$entrez [i] <- rownames(cnt_obj_filt) [dupes [i]]
}
means$toremove <- "FALSE"
i <- 1
c <- 1
while (i <= length(dupes)) {
  if (i==c) {
    c <- c+1
  } else if (means$entrez [i] == means$entrez [c]) {
    if(means$avg [i] > means$avg [c]) {
      means$toremove [c] <- T
    } else {
      means$toremove [i] <- T
    }
    i <- i+1
    c <- 1
  } else {
    c <- c+1
  }
}
dupes_to_remove <- means %>% filter(toremove == T)
cnt_obj_filt <- cnt_obj_filt [-dupes_to_remove$index,]
geneids_smbl <- mapIds(org.Hs.eg.db, keys = rownames(cnt_obj_filt), keytype = "ENTREZID", column = "SYMBOL", multiVals = "first") %>% 
  unlist() 
geneids_smbl <- data.frame(entrezid = names(geneids_smbl), symbol = unname(geneids_smbl))
```

```{r eggsea-kegg, eval=FALSE, fig.width=8, fig.height=6, eval=FALSE}
#Move onto EGSEA analysis
#Build gene set collection
gs.annots <-  buildIdx(entrezIDs = rownames(cnt_obj_filt), species = "human",
msigdb.gsets = "none")
#These pathways cause an error with report generation, so remove them prior to analysis
problematic_pathways <- c("Retrograde endocannabinoid signaling", "MicroRNAs in cancer") 
sel = which(names(gs.annots[["kegg"]]@original) %in% problematic_pathways) 
gs.annots[["kegg"]]@original = gs.annots[["kegg"]]@original[-sel] 
gs.annots[["kegg"]]@idx = gs.annots[["kegg"]]@idx[-sel] 
gs.annots[["kegg"]]@anno = gs.annots[["kegg"]]@anno[-sel, ]
#Perform the EGSEA analysis with voom built-in
gsa <-  egsea.cnt(counts = cnt_obj_filt, group = group_obj, design = design_matrix,
  contrasts = contrast_obj, gs.annots = gs.annots, symbolsMap = geneids_smbl,
  baseGSEAs = egsea.base(), sort.by = "p.adj", fdr.cutoff = 0.05, combineMethod = "wilkinson",
  num.threads = 4, report = TRUE, 
  report.dir = here::here(str_c(res_path,"all final comparisons")),
  verbose = T, display.top = 35)
#Now repeat just for WT analysis
contrast_obj_Wt <- makeContrasts(contrasts = names_for_contr[!str_detect(names_for_contr,"N")],
                                 levels = colnames(design_matrix))
gsa.wt <-  egsea.cnt(counts = cnt_obj_filt, group = group_obj, design = design_matrix,
  contrasts = contrast_obj_Wt, gs.annots = gs.annots, symbolsMap = geneids_smbl,
  baseGSEAs = egsea.base(), sort.by = "p.adj", fdr.cutoff = 0.05, combineMethod = "wilkinson",
  num.threads = 4, report = TRUE, 
  report.dir = here::here(str_c(res_path,"wt final comparisons")),
  verbose = T, display.top = 35)
gsets <- read_tsv(here::here(str_c(res_path, "wt final comparisons/ranked-gene-sets-base/ranked-kegg-gene-sets-WT180-WT0.txt")))
#png(here::here(str_c(res_path,"egsea WT180vsWT0 top pathways.png")), width = 8, height = 6, units = "in", res = 1080)
ggplot(dplyr::slice(gsets,1:15),
                     aes(x=-log10(p.adj), y = fct_reorder(str_wrap(GeneSet,30),-Rank), fill = as.character(direction))) +
                geom_bar(stat = "identity", color = "black", lwd = 1) +
                geom_vline(xintercept = -log10(0.05), linetype = "longdash", color = "grey10", lwd = 1.25) +
                labs(title = str_c("WT infection at 180min\nmost significant affected pathways"), y = element_blank(), 
                     fill = "Regulation direction") +
                scale_fill_manual(values = c("-1" = carto_pal(2, diverg_pal_choice) [1],
                                             "1" = carto_pal(2, diverg_pal_choice) [2]),
                                  labels = c("-1" = "Down-regulated","1" = "Up-regulated")) + 
                theme(title = element_text(size = 8),
                      legend.title = element_text(size = 12),
                      axis.title = element_text(size = 12),
                      axis.text = element_text(size = 12))
#dev.off()
```

```{r egsea-msigdb-h, eval=FALSE, fig.width=8, fig.height=6, eval=FALSE}
#Move onto EGSEA analysis with hallmark gene sets

#Build gene set collection
gs.annots.h <-  buildGMTIdx(geneIDs = rownames(cnt_obj_filt), species = "Human", 
                            gmt.file = here::here(str_c(data_path,"h.all.v7.5.1.entrez.gmt")))

gsa.h <-  egsea.cnt(counts = cnt_obj_filt, group = group_obj, design = design_matrix,
                    contrasts = contrast_obj_wt, gs.annots = gs.annots.h, symbolsMap = geneids_smbl,
                    baseGSEAs = egsea.base(), sort.by = "p.adj", fdr.cutoff = 0.05, combineMethod = "wilkinson",
                    num.threads = 4, report = TRUE, 
                    report.dir = here::here(str_c(res_path, "wt comparisons MSIGDB Hallmark")),
                    verbose = T, display.top = 35)
```

```{r egsea-msigdb-h-all-comparisons, fig.width=8, fig.height=6, eval=FALSE}
#Move onto EGSEA analysis with hallmark gene sets

#Build gene set collection
gs.annots.h <-  buildGMTIdx(geneIDs = rownames(cnt_obj_filt), species = "Human", 
                            gmt.file = here::here(str_c(data_path,"h.all.v7.5.1.entrez.gmt")))

gsa.h <-  egsea.cnt(counts = cnt_obj_filt, group = group_obj, design = design_matrix,
                    contrasts = contrast_obj, gs.annots = gs.annots.h, symbolsMap = geneids_smbl,
                    baseGSEAs = egsea.base(), sort.by = "p.adj", fdr.cutoff = 0.05, combineMethod = "wilkinson",
                    num.threads = 4, report = TRUE, 
                    report.dir = here::here(str_c(res_path, "all comparisons MSIGDB Hallmark")),
                    verbose = T, display.top = 35)
```
Plot the pathways from the comparison MSigDB
```{r prep-egsea-msigdb-h, fig.width=8, fig.height=5}
gsets_h_all_names <- list.files(path = here::here(str_c(res_path, "all comparisons MSIGDB Hallmark/ranked-gene-sets-base/")), pattern = "txt") %>% 
  str_subset("compare", negate = T)
#Read in the fies
gsets_h_list <- gsets_h_all_names %>% 
  set_names(.,nm = .) %>% 
  map(~read_tsv(here::here(str_c(res_path, "all comparisons MSIGDB Hallmark/ranked-gene-sets-base/",.x))))

#Rename the list for ease of operation
gsets_names <- gsets_h_all_names %>% 
  str_remove("ranked-gmtcustom-gene-sets-") %>% 
  str_remove("-WT0\\.txt")

#Display pathways with odd number of occurences (so they occur for one strain, but not the other) 
gsets_unq <- gsets_h_list %>% 
  set_names(gsets_names) %>%
  map(~filter(.x, p.adj < 0.05)) %>% 
  imap(~mutate(.x, Comparison = .y)) %>%
  reduce(bind_rows) %>% 
  select(GeneSet, Comparison) %>% 
  mutate(Strain = str_remove(Comparison, "[:digit:]+"),
         Time = str_remove(Comparison, "[:alpha:]+")) %>% 
  summarise(n_app = n(), .by = c("GeneSet", "Strain")) %>%  
  pivot_wider(names_from = "Strain", values_from = "n_app") %>% 
  mutate(WT = as.numeric(str_replace_na(WT, "0")),
         N = as.numeric(str_replace_na(N, "0"))) %>% 
  filter(WT != N)
#Filter to top n from WT180
top_n <- gsets_h_list$`ranked-gmtcustom-gene-sets-WT180-WT0.txt` %>% 
  slice_head(n=15) %>% 
  pull(var = "GeneSet")
gsets_h_filt <- gsets_h_list %>% 
  set_names(gsets_names) %>% 
  map(~filter(.x,GeneSet %in% gsets_unq$GeneSet)) %>% 
  map(select, GeneSet, direction, p.adj, Rank, avg.logfc) %>% 
  imap(~mutate(.x, Comparison = .y)) %>% 
  reduce(bind_rows) %>% 
  filter(p.adj < 0.05) %>% 
  mutate(Strain = str_remove(Comparison, "[:digit:]+"),
         Strain = fct_relevel(Strain, "WT"),
         Strain = fct_recode(Strain, 
                             "Wild-type EPEC" = "WT",
                             "T3SS-deficient EPEC" = "N"),
         Time = as.numeric(str_remove(Comparison, "[:alpha:]+")),
         direction = factor(direction),
         GeneSet = str_remove(GeneSet, "HALLMARK_"),
         GeneSet = str_replace_all(GeneSet, "_"," "),
         GeneSet = str_wrap(GeneSet, width = 25))
gsets_h_add_missing <- gsets_h_filt %>% 
  complete(GeneSet, Strain, Time)
```


```{r plot-egsea-msigdb-h, fig.width=8, fig.height=5}
#Plot the results
ggplot(gsets_h_filt, aes(x = factor(Time), y = GeneSet)) +
  geom_tile(aes(fill = direction), color = "black", alpha = 1) + 
  scale_shape_manual(values = set_names(c(24,25), nm = c("1","-1"))) +
  theme_bw(base_size = 11) + 
  theme(legend.title = element_text(face = "bold"),
        axis.text = element_text(face = "bold"),
        axis.title = element_text(face = "bold"),
        panel.grid.major = element_blank(),
        strip.text = element_text(face = "bold")) + 
  scale_fill_manual(values = c("-1" = carto_pal(2, diverg_pal_choice) [1],
                               "1" = carto_pal(2, diverg_pal_choice) [2]),
                    labels = c("-1" = "Down-regulated","1" = "Up-regulated")) +
  labs(x = "Time of Infection (min)", y = element_blank(), fill = "Regulation direction") +
  facet_wrap(~Strain, ncol = 2)
#ggsave(here::here(str_c(res_path,"egsea all top Hallmark pathways.png")), width = 8, height = 5, dpi = 1080)
#ggsave(here::here(str_c(res_path,"egsea all top Hallmark pathways.svg")), width = 8, height = 5)

ggplot(gsets_h_add_missing, aes(x = factor(Time), y = GeneSet)) +
  geom_tile(aes(fill = direction), color = "black", alpha = 1) + 
  scale_shape_manual(values = set_names(c(24,25), nm = c("1","-1"))) +
  theme_bw(base_size = 11) + 
  theme(legend.title = element_text(face = "bold"),
        axis.text = element_text(face = "bold"),
        axis.title = element_text(face = "bold"),
        panel.grid.major = element_blank(),
        strip.text = element_text(face = "bold")) + 
  scale_fill_manual(values = c("-1" = carto_pal(2, diverg_pal_choice) [1],
                               "1" = carto_pal(2, diverg_pal_choice) [2]),
                    labels = c("-1" = "Down-regulated","1" = "Up-regulated"),
                    na.value = "grey80") +
  labs(x = "Time of Infection (min)", y = element_blank(), fill = "Regulation direction") +
  facet_wrap(~Strain, ncol = 2)
#ggsave(here::here(str_c(res_path,"egsea all top Hallmark pathways with grey for missing.png")), width = 8, height = 5, dpi = 1080)
#ggsave(here::here(str_c(res_path,"egsea all top Hallmark pathways  with grey for missing.svg")), width = 8, height = 5)
    
```


```{r egsea-msigdb-h-wt-plot, eval=FALSE, fig.width=8, fig.height=6}
gsets_hall <- read_tsv(here::here(str_c(res_path, "wt comparisons MSIGDB Hallmark/ranked-gene-sets-base/ranked-gmtcustom-gene-sets-WT180-WT0.txt")))
#png(here::here(str_c(res_path,"egsea WT180vsWT0 top Hallmark pathways.png")), width = 8, height = 6, units = "in", res = 1080)
ggplot(dplyr::slice(gsets_hall,1:15),
       aes(x=-log10(p.adj), 
           y = fct_reorder(str_wrap(str_remove(str_replace_all(GeneSet,"_"," "),"HALLMARK "),20),-Rank), 
           fill = as.character(direction))) +
  geom_bar(stat = "identity", color = "black", lwd = 1) +
  geom_vline(xintercept = -log10(0.05), linetype = "longdash", color = "grey10", lwd = 1.25) +
  labs(title = str_c("WT 180 min Hallmark \nmost significant affected pathways"), y = element_blank(), 
       fill = "Regulation direction") +
  scale_fill_manual(values = c("-1" = carto_pal(2, diverg_pal_choice) [1],
                               "1" = carto_pal(2, diverg_pal_choice) [2]),
                    labels = c("-1" = "Down-regulated","1" = "Up-regulated")) + 
  theme(title = element_text(size = 8),
        legend.title = element_text(size = 12),
        axis.title = element_text(size = 12),
        axis.text = element_text(size = 10))
#dev.off()
```

Use clusterprofiler to compare themes between conditions
```{r cluster1, eval=FALSE}
res_wtvsn_list <- set_names(c("N180"),c("N180")) %>% 
  map(~results(dds, contrast = c("group",.x,"WT180"))) %>% 
  map(as_tibble,rownames = "gene_id") %>% 
  map(arrange,desc(log2FoldChange))
res_wtvsn_entrez <- res_wtvsn_list %>% 
  map(~mutate(.x, entrez_id = unlist(mapIds(org.Hs.eg.db, keys = str_replace(.x$gene_id, "\\.[1-9]([0-9])?", ""), keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first")))) %>% 
  map(~filter(.x, !is.na(entrez_id)))
#Get gene sets data sets
h_entrez <- read.gmt(here::here(str_c(data_path, "h.all.v7.5.1.entrez.gmt")))
c2_entrez <- read.gmt(here::here(str_c(data_path, "c2.cp.v2022.1.Hs.entrez.gmt")))
c2_kegg_entrez <- c2_entrez %>% 
  filter(str_detect(term, "KEGG"))
c2_react_entrez <- c2_entrez %>% 
  filter(str_detect(term, "REACTOME"))
res_wtvsn_entrez_list <- set_names(res_wtvsn_entrez[[1]]$log2FoldChange,res_wtvsn_entrez[[1]]$entrez_id) 
wtvsn_cp_list <- list(Hallmark = h_entrez,KEGG = c2_kegg_entrez,Reactome = c2_react_entrez) %>%  
  map(~GSEA(geneList = res_wtvsn_entrez_list, TERM2GENE = .x, pvalueCutoff = 0.05))
```


```{r cluster}
res_n_list <- set_names(c("N45","N90","N180"),c("N45","N90","N180")) %>% 
  map(~results(dds, contrast = c("group",.x,"WT0")))
res_n_list_lfc <- res_n_list %>% 
  map2(c("group_N45_vs_WT0","group_N90_vs_WT0","group_N180_vs_WT0"), 
       ~ lfcShrink(dds, coef=.y, type="apeglm", res = .x))

#Combine all shrunken results, and pull out DE genes
res_all_lfc_list <- list(WT45 = res45lfc, WT90 = res90lfc, WT180 = res180lfc) %>% 
  append(res_n_list_lfc) %>% 
  map(as_tibble,rownames = "gene_id")
res_select_lfc_list <- res_all_lfc_list %>% 
  map(dplyr::filter, padj < 0.05, abs(log2FoldChange)>0.5) %>% 
  map(arrange,padj) 
res_all_lfc_names <- res_select_lfc_list %>% 
  map(pull, gene_id)
res_all_lfc_entrez <- res_all_lfc_names %>% 
  map(~mapIds(org.Hs.eg.db, keys = str_replace(.x, "\\.[1-9]([0-9])?", ""), keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first")) %>% 
  map(unlist) %>% 
  map(~.x [!is.na(.x)])
universe <- mapIds(org.Hs.eg.db, keys = str_replace(rownames(res180), "\\.[1-9]([0-9])?", ""), keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first")
#Get gene sets data sets
h_entrez <- read.gmt(here::here(str_c(data_path, "h.all.v7.5.1.entrez.gmt")))
c2_entrez <- read.gmt(here::here(str_c(data_path, "c2.cp.v2022.1.Hs.entrez.gmt")))
c2_kegg_entrez <- c2_entrez %>% 
  filter(str_detect(term, "KEGG"))
c2_react_entrez <- c2_entrez %>% 
  filter(str_detect(term, "REACTOME"))
#Perform the clustering
ck_list <- list(Hallmark = h_entrez,KEGG = c2_kegg_entrez,Reactome = c2_react_entrez) %>% 
  map(~compareCluster(geneClusters = res_all_lfc_entrez, fun = "enricher",
                      TERM2GENE = .x, qvalueCutoff = 0.05,
                      pvalueCutoff = 0.05, universe = universe))
ck_readable_list <- ck_list %>% 
  map(setReadable,OrgDb = org.Hs.eg.db, keyType="ENTREZID")
```


```{r cluster-print, fig.width=8, fig.height=6}
biol_pw_df_list <- ck_readable_list %>% 
  map(as.data.frame) %>% 
  map(mutate, `Gene Ratio` = sapply(GeneRatio, function(x) eval(parse(text=x)))) %>% 
  map(mutate, ID = str_wrap(str_replace_all(ID, "_"," "),20))
#Print result as a table
#walk2(biol_pw_df_list, names(biol_pw_df_list),
 #     ~write_csv(.x, here::here(str_c(res_path,"tables/",.y," ora_theme_comparison.csv"))))
#Manipulate the reactome results to highlight discordances between response to WT and dEscN infection
react_res <- biol_pw_df_list [["Reactome"]] %>% 
  group_by(ID) %>% 
  nest() 
path_to_high <- c()
j <- 1
#Check which pathways are not present in symmetric conditions
for (pw in seq_along(react_res$ID)) {
  dummy_df <- react_res [[2]] [[pw]] 
  if(nrow(dummy_df) %% 2 != 0 & nrow(dummy_df) > 1) {
    path_to_high [j] <- pw
    j <- j +1
  } else if ("WT180" %in% dummy_df$Cluster & !("N180" %in% dummy_df$Cluster)) {
    path_to_high [j] <- pw
    j <- j +1
  }
}
react_res_select <- react_res [path_to_high,] %>% 
  unnest(cols = c(data)) %>% 
  mutate(ID = str_remove(ID, "REACTOME ")) 
biol_pw_df_list [["Reactome"]] <- react_res_select
#Plot the results
#pdf(here::here(str_c(res_path, "cluster compare themes.pdf")), width = 8, height = 7)
walk2(biol_pw_df_list, c(11,11,9),
     ~print(ggplot(.x, aes(x=Cluster, y = ID, color = p.adjust, size = `Gene Ratio`)) +
              geom_point() +
              scale_size_continuous(range = c(4,8)) +
              scale_color_carto_c(palette = cont_pal_choice, direction = -1) +
              theme_minimal(base_size = .y) +
              labs(x=element_blank(), y=element_blank()) +
              #scale_x_discrete(labels = c(WTN = bquote("WT vs" ~ Delta*"EscN EPEC"),
               #                           WTUI = "WT vs Uninfected",
                #                          NUI = bquote(Delta*"EscN EPEC vs Uninfected"))) +
              theme(panel.border = element_rect(fill = NA, color = "black"),
                    axis.text = element_text(color = "black"))))
#dev.off()
```

```{r euler-de, eval=FALSE}
all_de_genes_euler <- euler(res_all_lfc_names, shape = "ellipse")
#pdf(here::here(str_c(res_path,"overlap of de genes.pdf")), width = 8, height = 6)
plot(all_de_genes_euler, quantities = list("counts"))
#dev.off()
```



