---
title: "Final clean analysis of short-term infection RNA-seq"
output: 
  html_notebook: 
    fig_width: 6
    fig_height: 5
editor_options: 
  chunk_output_type: inline
---

```{r results='hide', message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(include = TRUE)
knitr::opts_chunk$set(fig.width = 6)
knitr::opts_chunk$set(fig.height = 5)
library(greekLetters)
library(RUVSeq)
library(EDASeq)
library(DESeq2)
library(PCAtools)
library('org.Hs.eg.db')
library(EGSEA)
library(limma)
library(pheatmap)
library(clusterProfiler)
library(tidyverse)
library(ggprism)
library(rcartocolor)
```

Define all helper functions and color choices
```{r include=TRUE}
theme_set(theme_prism(base_size = 14))
#Set color choices from CARTOcolors
discrete_pal_choice <- "Bold"
cont_pal_choice <- "BluYl"
diverg_pal_choice <- "Geyser"
#Helper functions
calc_relative <- function(x) x/sum(x) * 100
#Set up paths for data, and results
data_path <- "short_infection_human_reads/data/"
res_path <- "short_infection_human_reads/results/final clean figures/"
```

#Read in the count data and rename the columns to fit the sample names

```{r readorigstar}
counts <- read.table(here::here(str_c(data_path,"STAR_reads_updated.txt")),header = F, sep = "\t", stringsAsFactors = F)
#Next manipulate the table to remove redundant columns and only keep relevant columns (1 columns with gene names, and the rest with counts for reverse stranded library
counts_simple <- counts [,seq(from=4,to=96,by=4)]
colnames(counts_simple) <- c("4UIN", "4UIWT", "490WT", "490N", "4180N", "4180WT", "5UIN", "5UIWT", "545WT", "545N", "590WT", "590N", "5180N", "5180WT", "6UIN", "6UIWT", "645WT", "645N", "690WT", "690N", "6180N", "6180WT", "445N", "445WT")
#Prior analysis revealed that 6UIN is an outlier by PCA plot
counts_simple <- counts_simple %>% 
  dplyr::select(-`6UIN`)
counts_simple$Gene_ID <- counts$V1

filter <- apply(dplyr::select(counts_simple,-Gene_ID),1,function(x) mean(x)>5)
counts_simple <- counts_simple [filter,]

#write.table(counts_simple,file=here::here(str_c(res_path,"tables/only_reverse_strand_counts_cleaned.txt")),sep = "\t", row.names = F)
```

```{r vis-align-stats}
#Visualize alignment statistics
counts_stats <- counts_simple %>% 
  dplyr::slice(1:4) %>% 
  pivot_longer(-Gene_ID, values_to = "numbers", names_to = "mapping")
mapped <-  counts_simple %>%
  dplyr::select(-Gene_ID) %>%
  summarise(across(everything(), sum)) %>% 
  pivot_longer(everything(),values_to = "numbers", names_to = "mapping") %>% 
  mutate(Gene_ID = "Mapped") %>% 
  bind_rows(counts_stats) %>% 
  mutate(Gene_ID = fct_relevel(str_to_title(str_replace(Gene_ID, "N_","")), "Mapped"))

#Raw counts stats
raw_count_plot <- ggplot(mapped, aes(x=mapping, y=numbers, fill = Gene_ID)) +
  geom_bar(stat = "identity") + scale_fill_carto_d(palette = discrete_pal_choice) +
  labs(x = "", y = "Raw counts") +theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6)) + 
  guides(fill = guide_legend(title = "Mapping Status"))
#Relative count stats
rel_map <- mapped %>% 
  pivot_wider(Gene_ID, names_from = mapping, values_from = numbers) %>%  
  mutate(across(-Gene_ID, calc_relative)) %>% 
  pivot_longer(-Gene_ID,values_to = "numbers", names_to = "mapping")
rel_count_plot <-ggplot(rel_map, aes(x=mapping, y=numbers, fill = Gene_ID)) +
  geom_bar(stat = "identity") + scale_fill_carto_d(palette = discrete_pal_choice) +
  labs(x = "", y = "Proportion of counts") +
  theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6)) + 
  guides(fill = guide_legend(title = "Mapping Status"))
#pdf(file = here::here(str_c(res_path,"mapping stats.pdf")), width = 8, height = 6)
raw_count_plot
rel_count_plot
#dev.off()
```

Calculate statistics of types of transcripts mapped

```{r map-by-type}
geneids <- mapIds(org.Hs.eg.db, keys = str_replace_all(counts_simple$Gene_ID [-1:-4], "\\.[1-9]([0-9])?", ""), 
                  keytype = "ENSEMBL", column = "GENETYPE", multiVals = "first") %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "Gene_ID")
c_type <- counts_simple %>% 
  dplyr::slice(-1:-4) %>% 
  mutate(Gene_ID = str_replace_all(Gene_ID, "\\.[1-9]([0-9])?", "")) %>% 
  left_join(geneids) %>% 
  dplyr::select(-Gene_ID) %>%
  rename_with(function(x) "Type",last_col()) %>% 
  group_by(Type) %>% 
  summarise(across(everything(),sum)) %>% 
  mutate(across(-Type, calc_relative)) %>% 
  pivot_longer(-Type, values_to = "numbers", names_to = "mapping") %>% 
  mutate(Type = str_replace_na(Type, replacement = "other"))
rel_type_plot <- ggplot(c_type, aes(x=mapping, y=numbers, fill = fct_reorder(Type,numbers, .desc = T))) +
  geom_bar(stat = "identity") + scale_fill_carto_d(palette = discrete_pal_choice) +
  labs(x = "", y = "Proportion of genes") +
  theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6)) + 
  guides(fill = guide_legend(title = "Gene type"))
#pdf(file = here::here(str_c(res_path,"mapping stats by gene type.pdf")), width = 8, height = 6)
rel_type_plot
#dev.off()
```


Apply EDAseq normalization to the data

```{r edaseq}
#Use EDAseq to normalize the libraries
c_mat <- counts_simple %>% 
                     dplyr::select(-dplyr::last_col()) %>% 
                     dplyr::slice(-1:-4) %>% 
                     as.matrix()
rownames(c_mat) <- counts_simple$Gene_ID [-1:-4]
pData <- data.frame(genotype=vector(mode = "logical", length = ncol(c_mat)),
                    timepoint = vector(mode = "logical", length = ncol(c_mat)),
                    samples = colnames(c_mat),
                    row.names = colnames(c_mat))
pData <- pData %>% 
  mutate(genotype = case_when(
    str_detect(samples, "WT") ~ "WT",
    str_detect(samples, "UIN") ~ "WT",
    str_detect(samples, "N") ~ "N"
  ), timepoint = case_when(
    str_detect(samples, "UI") ~ 0,
    str_detect(samples, "45") ~ 45,
    str_detect(samples, "90") ~ 90,
    str_detect(samples, "180") ~ 180,
  )) %>% dplyr::select(-samples)

data <- newSeqExpressionSet(counts=c_mat,
                            phenoData=pData)
#Normalize the data next
dataNorm <- betweenLaneNormalization(data, which="upper") 
#Plot the counts
boxplot(data, main = "Unnormalized counts", col = "dodgerblue3")
boxplot(dataNorm, main = "Normalized counts", col = "dodgerblue3")
#For DE we will create this analysis with offset values
dataOffset <- betweenLaneNormalization(data,
                                       which="upper",offset=TRUE)
```
Next step is to use normalized counts in the DESEq2 analysis of differential expression
```{r deseq, fig.width=8, fig.height=6}
#Prepare the DESEq2 object and apply EDASeq normalization factors
pData$group <- factor(paste0(pData$genotype,pData$timepoint))
pData <- pData %>% 
  mutate(genotype = case_when(
    timepoint == 0 ~ "Uninfected",
    TRUE ~ genotype
  ))
dds <- DESeqDataSetFromMatrix(countData = counts(dataOffset),
                              colData = pData,
                              design = ~ group)
dds$group <- relevel(dds$group, ref= "WT0")
normFactors <- exp(-1 * offst(dataOffset))
normFactors <- normFactors / exp(rowMeans(log(normFactors)))
normalizationFactors(dds) <- normFactors

#Perform the DESEq2 analysis
dds <- DESeq(dds)
rld <- rlog(dds)

normal_PCA <- plotPCA(rld, returnData = T, intgroup = "group") %>% 
  mutate(timepoint = as.numeric(str_remove(group, "[:alpha:]+")),
         genotype = ifelse(timepoint == 0, "Uninfected", str_remove(group, "[:digit:]+"))) %>% 
  mutate(genotype = fct_relevel(genotype, "Uninfected", "WT", "N"))
  
#Plot the PCA graphs of the normalized counts
#png(here::here(str_c(res_path,"PCA plot of normalized counts.png")), width = 8, height = 6, units = "in", res = 1080)
ggplot(data = normal_PCA, aes(shape = genotype)) +
  scale_color_carto_c(palette = cont_pal_choice, breaks = c(0,45,90,180)) +
  geom_point(mapping = aes(x = PC1, y = PC2, color = timepoint), size = 5) +
  labs(title = "EDAseq normalized DESeq2 PCA plot \nnot including the outlier",
       x = "PC1: 39% variance",
       y = "PC2: 21% variance") +
  guides(color = guide_legend(title = "Time of infection (min)"),
         shape = guide_legend(title = "Infection status")) +
  scale_shape_manual(labels = c("WT" = "Wild-type EPEC",
                                "N" = paste0(greeks("Delta"),"EscN EPEC")),
              values = c("Uninfected" = 15, "WT" = 16, "N" = 17)) +
  theme(legend.title = element_text())
#dev.off()
```
Identify differentially expressed genes
```{r deseq-res}
res45 <- results(dds, contrast = c("group","WT45", "WT0"))
res90 <- results(dds, contrast = c("group","WT90", "WT0"))
res180 <- results(dds, contrast = c("group","WT180", "WT0"))
time_res <- list(res45,res90,res180) %>% 
  map(as.data.frame) %>% 
  map2(c(".45",".90",".180"), ~rename_with(.x, .fn = function(colname) str_c(colname,.y))) %>% 
  map(rownames_to_column, var = "geneid") %>% 
  reduce(left_join)
#write.csv(time_res, here::here(str_c(res_path,"tables/EDAseq_normalized results comparing by time WT infection.csv")), row.names = F)
```


```{r deseq-res-wt-n}
res90vsn <- results(dds, contrast = c("group","WT90", "N90"))
res180vsn <- results(dds, contrast = c("group","WT180", "N180")) #none showed significant differences
summary(res90vsn)
summary(res180vsn)
```
Analyze loadings of the principal components in the data
```{r pcatools, fig.width=9, fig.height=6}
#Create a mapping for ID to symbol
geneids <- mapIds(org.Hs.eg.db, keys = str_replace_all(rownames(counts(dds)), "\\.[1-9]([0-9])?", ""), 
                  keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")

cnts <- assay(rld) 
#Confirm that that the mapping is correct
all(str_replace_all(rownames(counts(dds)), "\\.[1-9]([0-9])?", "") == names(geneids))
#Rename rows that had a symbol name
newnames <- ifelse(is.na(geneids) | duplicated(geneids),
    names(geneids), geneids)
rownames(cnts) <- newnames
p <- pca(cnts,colData(dds), removeVar = 0.1)
#pdf(here::here(str_c(res_path,"PCA tools loadings of PCA.pdf")), width = 9, height = 6)
plotloadings(p,  rangeRetain = 0.01,
    labSize = 2.5, col = carto_pal(3,diverg_pal_choice) ,
    caption = "Top 1% variables") + theme(legend.title = element_text())
#dev.off()
```
Make volcano plots of differentially expressed genes
```{r volcano}
library(EnhancedVolcano)
library('org.Hs.eg.db')
#shrink results
res90lfc <- lfcShrink(dds, coef="group_WT90_vs_WT0", type="apeglm", res = res90)
res180lfc <- lfcShrink(dds, coef="group_WT180_vs_WT0", type="apeglm", res = res180)

#Create a mapping for ID to symbol
geneids <- mapIds(org.Hs.eg.db, keys = str_replace_all(rownames(res180lfc), "\\.[1-9]([0-9])?", ""), 
                  keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first") %>% 
  as.data.frame() %>% 
  bind_cols(rownames(res180lfc)) %>% 
  dplyr::select("symbol" = 1, "rowname" = `...2`)
#Get nice names or keep ensembl IDs
naOrDup <- is.na(geneids$symbol) | duplicated(geneids$symbol)
geneids$gene_name <- ifelse(naOrDup, rownames(geneids), geneids$symbol)
#identify top20 genes for each comparison
id_top <- function(x) {
  x %>% 
  as.data.frame() %>% 
  arrange(padj) %>% 
  dplyr::slice(1:20) %>% 
  rownames_to_column(var = "rowname") %>% 
  left_join(geneids) }
top_list <- list("WT90" = res90lfc,"WT180" = res180lfc) %>% 
  map(id_top)
results_list <- list(res90lfc,res180lfc)
#Replace the rownames with gene names
for (i in 1:2) {rownames(results_list [[i]]) <- geneids$gene_name}
#pdf(here::here("final clean figures/Volcano plots.pdf"), width = 8, height = 6.6)
plotMA(res90, main = "WT infection 90 min vs 0min \nResults without shrinking")
plotMA(res90lfc, main = "WT infection 90 min vs 0min \nResults after apeglm shrinking")
plotMA(res180, main = "WT infection 180 min vs 0min \nResults without shrinking")
plotMA(res180lfc, main = "WT infection 180 min vs 0min \nResults after apeglm shrinking")

volcano_with_labels <- function(res_obj,select_lab = "",title) {
  EnhancedVolcano(res_obj,
    lab = rownames(res_obj),
    selectLab = select_lab,
    x = 'log2FoldChange',
    y = 'pvalue',
    title = title,
    pointSize = 3,
    col = c("darkgrey","red3","dodgerblue2","forestgreen"),
    legendPosition = 'bottom',
    subtitle = NULL,
    pCutoff = 0.05,
    FCcutoff = 0.6,
    labSize = 4.0,
                boxedLabels = T, labCol = 'black',
               labFace = 'bold',
                drawConnectors = T, widthConnectors = 1.0,
               colConnectors = 'black')
}
titles <- list("Differentially expressed genes at 1.5hr of infection",
               "Differentially expressed genes at 3hr of infection")
for (i in 1:2) {
  volcano_with_labels(results_list[[i]],title = titles [[i]]) %>% print()
  volcano_with_labels(results_list[[i]],select_lab = top_list [[i]]$gene_name,title = titles [[i]]) %>% print()
}
#dev.off()
```

Create a heatmap of top differentially expressed genes
```{r heatmap, eval=FALSE}
library(pheatmap)
heat_mat <- assay(rld)
#Create a mapping for ID to symbol
geneids <- mapIds(org.Hs.eg.db, keys = str_replace_all(rownames(heat_mat), "\\.[1-9]([0-9])?", ""), 
                  keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first") %>% 
  as.data.frame() %>% 
  bind_cols(rownames(heat_mat)) %>% 
  dplyr::select("symbol" = 1, "rowname" = `...2`)
#Get nice names or keep ensembl IDs
naOrDup <- is.na(geneids$symbol) | duplicated(geneids$symbol)
geneids$gene_name <- ifelse(naOrDup, rownames(geneids), geneids$symbol)
rownames(heat_mat) <- geneids$gene_name
topVarGenes <- head(order(-rowVars(heat_mat)),20)
mat <- heat_mat [ topVarGenes, ]
mat <- mat - rowMeans(mat)
#For some reason fct_recode did not want to work the delta sign, so had to work around it
df <- as.data.frame(colData(rld)) %>%  
  mutate(`Infection status` = fct_recode(genotype, "T3SS-deficient EPEC" = "N",
                                            "Wild-type EPEC" = "WT","Uninfected" = "UI"), .keep = 'unused') %>% 
  mutate(timepoint = fct_recode(as.factor(timepoint), "Uninfected" = "0",
                                "90 min infection" = "90","180 min infection" = "180"))

pheatmap(mat, annotation_col=df)
```


Apply EGSEA to differentially expressed data

```{r eggsea-prep}
library(limma)
#remove intercept since this is a factor. Intercept or no intercept does not make a difference
#That is what the 0+ means in the formula
design_matrix <- model.matrix(~0+pData$group) 
colnames(design_matrix) <- colnames(design_matrix) %>% str_replace("pData\\$group","") 
names_for_contr <- resultsNames(dds) [-1] %>%
  str_replace_all("group_","") %>% 
  str_replace_all("_vs_","-")
contrast_obj <- makeContrasts(contrasts = names_for_contr, levels = colnames(design_matrix))
group_obj <- as.factor(pData$group)
cnt_obj <- dataNorm@assayData$normalizedCounts
#finally prepare the geneID mapping tool
library('org.Hs.eg.db')
geneids <- mapIds(org.Hs.eg.db, keys = str_replace(rownames(cnt_obj), "\\.[1-9]([0-9])?", ""), keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first")

#Replace the rownames of the count matrix with ENTREZ ID
rownames(cnt_obj) <- unname(geneids)
#Remove NAs from the matrix (not every gene had an NCBI match)
cnt_obj_filt <- cnt_obj [!is.na(rownames(cnt_obj)),] 
#Deal with the remaining duplicates
dupes <-  c(which(duplicated(rownames(cnt_obj_filt)) | duplicated(rownames(cnt_obj_filt),fromLast = T)))
means <- data.frame(matrix(nrow = length(dupes))) %>% mutate(index = dupes)
#Select the duplicate with higher gene count
for (i in 1:length(dupes)) {
  means$avg [i] <- mean(cnt_obj_filt[dupes[i],])
  means$entrez [i] <- rownames(cnt_obj_filt) [dupes [i]]
}
means$toremove <- "FALSE"
i <- 1
c <- 1
while (i <= length(dupes)) {
  if (i==c) {
    c <- c+1
  } else if (means$entrez [i] == means$entrez [c]) {
    if(means$avg [i] > means$avg [c]) {
      means$toremove [c] <- T
    } else {
      means$toremove [i] <- T
    }
    i <- i+1
    c <- 1
  } else {
    c <- c+1
  }
}
dupes_to_remove <- means %>% filter(toremove == T)
cnt_obj_filt <- cnt_obj_filt [-dupes_to_remove$index,]
geneids_smbl <- mapIds(org.Hs.eg.db, keys = rownames(cnt_obj_filt), keytype = "ENTREZID", column = "SYMBOL", multiVals = "first") %>% 
  unlist() 
geneids_smbl <- data.frame(entrezid = names(geneids_smbl), symbol = unname(geneids_smbl))
```
Use clusterprofiler to compare themes between conditions
```{r cluster}
library(clusterProfiler)
wt180_names <- res180 %>% 
  as_tibble(rownames = "gene_id") %>% 
  filter(padj < 0.1) %>% 
  dplyr::arrange(padj) %>% 
  pull(gene_id)
n180_names <- results(dds, contrast = c("group","N180", "WT0")) %>% 
  as_tibble(rownames = "gene_id") %>% 
  filter(padj < 0.1) %>% 
  dplyr::arrange(padj) %>% 
  pull(gene_id)
wt180_entrez <- mapIds(org.Hs.eg.db, keys = str_replace(wt180_names, "\\.[1-9]([0-9])?", ""), keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first") %>% 
  unlist()
wt180_entrez <- wt180_entrez [!is.na(wt180_entrez)|!duplicated(wt180_entrez)]
n180_entrez <- mapIds(org.Hs.eg.db, keys = str_replace(n180_names, "\\.[1-9]([0-9])?", ""), keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first") %>% 
  unlist()
n180_entrez <- n180_entrez [!is.na(n180_entrez)|!duplicated(n180_entrez)]
gc <- list("WTvsUI" = wt180_entrez,"EscNvsUI" = n180_entrez)
ck <- compareCluster(geneCluster = gc, fun = enrichKEGG)
ck <- setReadable(ck, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
png(here::here("final clean figures/ClusterEnricher Compare WTvsN enriched pathways.png"), width = 12, height = 9,
    units = "in",res = 720)
dotplot(ck, showCategory = 13, font.size = 16)
dev.off()
```




```{r eggsea, eval=FALSE}
library(EGSEA)
#Move onto EGSEA analysis
#Build gene set collection
gs.annots <-  buildIdx(entrezIDs = rownames(cnt_obj_filt), species = "human",
msigdb.gsets = "none")
#These pathways cause an error with report generation, so remove them prior to analysis
problematic_pathways <- c("Retrograde endocannabinoid signaling", "MicroRNAs in cancer") 
sel = which(names(gs.annots[["kegg"]]@original) %in% problematic_pathways) 
gs.annots[["kegg"]]@original = gs.annots[["kegg"]]@original[-sel] 
gs.annots[["kegg"]]@idx = gs.annots[["kegg"]]@idx[-sel] 
gs.annots[["kegg"]]@anno = gs.annots[["kegg"]]@anno[-sel, ]
#Perform the EGSEA analysis with voom built-in
gsa <-  egsea.cnt(counts = cnt_obj_filt, group = group_obj, design = design_matrix,
  contrasts = contrast_obj, gs.annots = gs.annots, symbolsMap = geneids_smbl,
  baseGSEAs = egsea.base(), sort.by = "p.adj", fdr.cutoff = 0.1, combineMethod = "wilkinson",
  num.threads = 4, report = TRUE, report.dir = here::here("final clean figures/all final comparisons"),
  verbose = T, display.top = 35)
#Now repeat just for WT analysis
contrast_obj_Wt <- makeContrasts(contrasts = names_for_contr[!str_detect(names_for_contr,"N")],
                                 levels = colnames(design_matrix))
gsa.wt <-  egsea.cnt(counts = cnt_obj_filt, group = group_obj, design = design_matrix,
  contrasts = contrast_obj_Wt, gs.annots = gs.annots, symbolsMap = geneids_smbl,
  baseGSEAs = egsea.base(), sort.by = "p.adj", fdr.cutoff = 0.1, combineMethod = "wilkinson",
  num.threads = 4, report = TRUE, report.dir = here::here("final clean figures/wt final comparisons"),
  verbose = T, display.top = 35)
gsets <- read_tsv(here::here("final clean figures/wt final comparisons/ranked-gene-sets-base/ranked-kegg-gene-sets-compare.txt"))
#png(here::here("final clean figures/egsea WTvsWT0 top pathways.png"), width = 7, height = 9, units = "in", res = 720)
ggplot(dplyr::slice(gsets,1:26),
       aes(x=-log10(p.adj), y = fct_reorder(str_wrap(GeneSet,30),-Rank), fill = as.character(direction))) +
  geom_bar(stat = "identity", color = "black", lwd = 1) +
  geom_vline(xintercept = 1, linetype = "longdash", color = "red2", lwd = 1) +
  labs(y = '', fill = "Regulation direction") +
  scale_fill_manual(values = c("-1" = "deepskyblue3","1" = "coral", "0" = "mediumpurple3"),
                    labels = c("-1" = "Down-regulated","1" = "Up-regulated", "0" = "Neutral")) +
  theme_minimal(base_size = 14) + theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank(),
                                        axis.text.y = element_text(hjust = 1))
#dev.off()
```

```{r egsea-msigdb-h, eval=FALSE}
#Move onto EGSEA analysis with hallmark gene sets

#Build gene set collection
gs.annots.h <-  buildGMTIdx(geneIDs = rownames(cnt_obj_filt), species = "Human", gmt.file = here::here("h.all.v7.5.1.entrez.gmt"))

gsa.h <-  egsea.cnt(counts = cnt_obj_filt, group = group_obj, design = design_matrix,
  contrasts = contrast_obj_Wt, gs.annots = gs.annots.h, symbolsMap = geneids_smbl,
 baseGSEAs = egsea.base(), sort.by = "p.adj", fdr.cutoff = 0.1, combineMethod = "wilkinson",
  num.threads = 4, report = TRUE, report.dir = here::here("final clean figures/wt comparisons MSIGDB Hallmark"),
  verbose = T, display.top = 35)
```


