---
title: "Finalized script from mirDeep2 analysis of long infection"
output: 
  html_notebook: 
    fig_width: 6
    fig_height: 5
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(include = TRUE)
library(biomaRt)
library(limma)
library(greekLetters)
library(RUVSeq)
library(EDASeq)
library(DESeq2)
library(ggprism)
library(eulerr)
library(tidyverse)
library(rcartocolor)
theme_set(theme_prism(base_size = 14))
```

Prepare the data for analysis
```{r echo=TRUE, results = 'markup', include=TRUE}
#analyzing miRNA-seq library prep for TC7 infection for 9 hours
counts <- read.table(here::here("long_infection_human_reads/miRNA_library/data/all_reads_mirDEEP2"), 
                     header = F, sep = "\t", stringsAsFactors = F)
#Next manipulate the table to remove redundant columns and only keep relevant columns (1 columns with gene names, and the rest with counts for reverse stranded library)
counts_simple <- counts %>% dplyr::select(Mature_miRNA = V1, precursor = V3, UI92 = V2, UI93 = V8, UI96 = V14, N92 = V20, N93 = V26, WT96 = V32, WT91 = V38, WT94 = V44, N95 = V50)
filter <- apply(counts_simple [,3:11],1,function(x) mean(x)>5)
counts_simple <- counts_simple [filter,]
```

Apply EDAseq normalization to the data

```{r edaseq}
#Use EDAseq to normalize the libraries
c_mat <- counts_simple %>% 
                     dplyr::select(-where(is.character)) %>% 
                     as.matrix()
rownames(c_mat) <- paste0(counts_simple$Mature_miRNA,"_",counts_simple$precursor)
pData <- data.frame(genotype=vector(mode = "logical", length = ncol(c_mat)),
                    timepoint = vector(mode = "logical", length = ncol(c_mat)),
                    samples = colnames(c_mat),
                    row.names = colnames(c_mat))
pData <- pData %>% 
  mutate(genotype = case_when(
    str_detect(samples, "WT") ~ "WT",
    str_detect(samples, "UI") ~ "UI",
    str_detect(samples, "N") ~ "N"
  ), timepoint = 9) %>%
  dplyr::select(-samples) %>% 
  mutate(graph_cols  = case_when(
    genotype == "UI" ~ carto_pal(12,discrete_pal_choice) [10],
    genotype == "WT" ~ carto_pal(12,discrete_pal_choice) [2],
    genotype == "N" ~ carto_pal(12,discrete_pal_choice) [1]
  ))

data <- newSeqExpressionSet(counts=c_mat,
                            phenoData=pData)
#Normalize the data next
dataNorm <- betweenLaneNormalization(data, which="full")
#For DE we will create this analysis with offset values
dataOffset <- betweenLaneNormalization(data,
                                       which="full",offset=TRUE)
#Plot the counts
plotPCA(data, col = pData$graph_cols, main = "Unnormalized counts", cex.axis = 1.5, cex.lab = 1.5)
plotPCA(dataOffset, col = pData$graph_cols, main = "Normalized counts", cex.axis = 1.5, cex.lab = 1.5)
plotRLE(data, col = pData$graph_cols, main = "RLE Unnormalized counts", cex.axis = 1.5)
plotRLE(dataOffset, col = pData$graph_cols, main = "RLE Normalized counts", cex.axis = 1.5)
boxplot(data, main = "Unnormalized counts", col = pData$graph_cols, cex.axis = 1.5)
boxplot(dataNorm, main = "Normalized counts", col = pData$graph_cols, cex.axis = 1.5)
```

Next step is to use normalized counts in the DESEq2 analysis of differential expression
```{r deseq}
#Prepare the DESEq2 object
pData$group <- factor(paste0(pData$genotype,pData$timepoint))
dds <- DESeqDataSetFromMatrix(countData = counts(dataOffset),
                              colData = pData,
                              design = ~ group)
dds$group <- relevel(dds$group, ref= "UI9")
normFactors <- exp(-1 * offst(dataOffset))
normFactors <- normFactors / exp(rowMeans(log(normFactors)))
normalizationFactors(dds) <- normFactors
#Perform the DESEq2 analysis
dds <- DESeq(dds)
rld <- varianceStabilizingTransformation(dds)

normal_PCA <- plotPCA(rld, returnData = T, intgroup = "group")
#Plot the PCA graphs of the normalized counts
#png(here::here("long_infection_human_reads/miRNA_library/results/final clean figures/PCA plot of normalized counts legend bottom.png"), width = 6, height = 5, units = "in", res = 1080)
ggplot(data = normal_PCA, aes(shape = pData$genotype, color = pData$genotype)) +
  scale_color_manual(labels = c("N" = paste0(greeks("Delta"),"EscN EPEC"),
                                "WT" = "Wild-type EPEC", "UI" = "Uninfected"),
                     values = c("UI" = carto_pal(12,discrete_pal_choice) [10],
                                "WT" = carto_pal(12,discrete_pal_choice) [2], 
                                "N" = carto_pal(12,discrete_pal_choice) [1])) +
  geom_point(mapping = aes(x = PC1, y = PC2), size = 5) +
  labs(title = "EDAseq-norm PCA plot", shape = "Infection", color = "Infection",
       x = "PC1: 33% variance", y = "PC2: 22% variance") +
  scale_shape_manual(labels = c("N" = paste0(greeks("Delta"),"EscN EPEC"),
                                "WT" = "Wild-type EPEC", "UI" = "Uninfected"),
                     values = c("UI" = 15, "WT" = 16, "N" = 17)) +
  theme(legend.position = "bottom", legend.title = element_text())
#dev.off()
```

Identify differentially expressed genes
```{r deseq-res-1}
resWTvsUI <- results(dds, contrast = c("group","WT9", "UI9"))
resWTvsN <- results(dds, contrast = c("group","WT9", "N9"))
resNvsUI <- results(dds, contrast = c("group","N9", "UI9"))
res <- list(resWTvsUI,resWTvsN,resNvsUI) %>% 
  map(as.data.frame) %>% 
  map(rownames_to_column, var = "rowname") %>% 
  purrr::reduce(left_join, by = "rowname") %>% 
  rename_with(function(x) str_replace(x,"$",".NvsUI"),
              .cols = !matches("\\.[xy]")) %>% 
  rename_with(function(x) ifelse(str_detect(x,"\\.x"),
                                 str_replace(x,"\\.x",".WTvsUI"),
                                 str_replace(x,"\\.y",".WTvsN")),
              .cols = matches("\\.[xy]")) %>% 
  rename_with(function(x) "Gene ID",1) %>% 
  separate(`Gene ID`, into = c("Mature miRNA", "Precursor miRNA"), sep = "_")
walk(list(resWTvsUI,resWTvsN,resNvsUI),summary)
#write.table(res, here::here("final clean figures/normalized results.txt"),
 #          sep = "\t", row.names = F)
```
Next step is to determine overwhelming presence of miRNA (some can make up 50% of the data)
```{r abund-mirna}
calc_relative <- function(x) x/sum(x)*100
rel_abund <- counts(dds) %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "rowname") %>% 
  mutate(across(!where(is.character), calc_relative)) %>%  
  arrange(desc(UI92),desc(UI93)) %>% 
  mutate(top_names = c(rowname [1:5],rep("other",nrow(.)-5))) %>% 
  pivot_longer(-c("rowname","top_names"), names_to = "mapping", values_to = "numbers")
#remove the highest abundant miRNA to see how the analysis falls
rel_abund_no_high <- counts(dds) %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "rowname") %>% 
  filter(rowname != "hsa-miR-192-5p_hsa-mir-192" & rowname != "hsa-miR-215-5p_hsa-mir-215") %>%
  mutate(across(!where(is.character), calc_relative)) %>%  
  arrange(desc(UI92),desc(UI93)) %>% 
  mutate(top_names = c(rowname [1:5],rep("other",nrow(.)-5))) %>% 
  pivot_longer(-c("rowname","top_names"), names_to = "mapping", values_to = "numbers")
#pdf(here::here("final clean figures/miRNA mapping stats by miRNA identity1.pdf"),width = 12,height = 10)
rel_list <- list(rel_abund,rel_abund_no_high)
rel_point <- rel_abund %>% 
  group_by(top_names, mapping) %>% 
  summarise(numbers = mean(numbers)) %>% 
  mutate(condition = case_when(
    str_detect(mapping,"WT") ~ "WT",
    str_detect(mapping,"N") ~ "N",
    str_detect(mapping,"UI") ~ "UI",
  ))
for (i in 1:2) {
  p <- ggplot(rel_list [[i]], aes(x=mapping, y=numbers, fill = fct_reorder(top_names,numbers, .desc = T))) +
    geom_bar(stat = "identity") + scale_fill_brewer(name = "Aligned to miRNA",palette = "Accent") +
    labs(x = "", y = "Proportion of counts") +
    theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6))
  print(p)
}

ggplot(rel_point, aes(x = mapping, y = fct_reorder(top_names,numbers, .desc = T), size = numbers, fill = condition)) + 
      geom_point(alpha = 0.75, shape = 21) + 
       scale_size_continuous(range = c(1,20), breaks = c(5,10,20,35)) + 
       labs( x= "", y = "", size = "Relative Abundance (%)", fill = "Condition")  + 
       scale_fill_manual(name = "Infection status",
                    labels = c("N" = paste0(greeks("Delta"),"EscN EPEC"),
                                "WT" = "Wild-type EPEC", "UI" = "Uninfected"),
                     values = c("UI" = "coral","WT" = "darkblue", "N" = "forestgreen"))
#dev.off()
```

Remove the two highest abundant miRNA from analysis and repeat analysis
```{r deseq-1}
data.n <- newSeqExpressionSet(counts=c_mat [rownames(c_mat) != "hsa-miR-192-5p_hsa-mir-192" 
                                          & rownames(c_mat) != "hsa-miR-215-5p_hsa-mir-215",],
                            phenoData=pData)
#Normalize the data next
dataNorm.n <- betweenLaneNormalization(data.n, which="full") 
#Plot the counts 
#pdf(here::here("final clean figures/Normalization effect on the library.pdf"), width = 12, height = 10)
boxplot(data.n, main = "Unnormalized counts", col = "dodgerblue3")
boxplot(dataNorm.n, main = "Normalized counts", col = "dodgerblue3")
#dev.off()
#For DE we will create this analysis with offset values
dataOffset.n <- betweenLaneNormalization(data.n,
                                       which="full",offset=TRUE)
#Prepare the DESEq2 object and apply EDASeq normalization factors
pData$group <- factor(paste0(pData$genotype,pData$timepoint))
dds.new <- DESeqDataSetFromMatrix(countData = counts(dataOffset.n),
                              colData = pData,
                              design = ~ group)
dds.new$group <- relevel(dds.new$group, ref= "UI9")
normFactors <- exp(-1 * offst(dataOffset))
normFactors <- normFactors / exp(rowMeans(log(normFactors)))
normalizationFactors(dds) <- normFactors
#Perform the DESEq2 analysis
dds.new <- DESeq(dds.new)
```
Identify differentially expressed genes
```{r deseq-res}
resWTvsUI.new <- results(dds.new, contrast = c("group","WT9", "UI9"))
resWTvsN.new <- results(dds.new, contrast = c("group","WT9", "N9"))
resNvsUI.new <- results(dds.new, contrast = c("group","N9", "UI9"))
summary(resWTvsUI.new)
summary(resWTvsN.new)
#Removing top 2 miRNA does not affect differential expression analysis, so keep them in
```

Perform PCA tools analysis of the data
```{r pcatools}
cnts <- assay(rld)
#Confirm that that the mapping is correct
rownames(cnts) <- rownames(cnts) %>% str_replace_all("hsa-","")
p <- pca(cnts,colData(dds), removeVar = 0.1)
p$metadata$genotype <- p$metadata$genotype %>% as.factor() %>% fct_relevel("UI")
#pdf(here::here("final clean figures/PCA tools analysis of PCA.pdf"), width = 12, height = 10)
biplot(p, colby = "genotype", shape = "genotype", showLoadings = T,gridlines.major = FALSE)
biplot(p, colby = "genotype", shape = "genotype",gridlines.major = FALSE)
plotloadings(p,  rangeRetain = 0.01,
    labSize = 4.0,
    caption = "Top 1% variables")
eigencorplot(p,
             components = getComponents(p,c(1:9)),
             metavars = c('group','genotype'))
eigencorplot(p,
             components = getComponents(p,1:9),
    metavars = c('group','genotype'),
    col = c('white', 'cornsilk1', 'gold', 'forestgreen', 'darkgreen'),
    cexCorval = 2,
    cexLabX = 1.8,
    cexLabY = 1.8,
    cexLabColKey = 1.8,
    fontCorval = 2,
    posLab = 'bottomleft',
    rotLabX = 45,
    scale = TRUE,
    main = bquote(Principal ~ component ~ Pearson ~ r^2 ~ correlates),
    plotRsquared = TRUE,
    corFUN = 'pearson',
    corUSE = 'pairwise.complete.obs',
    corMultipleTestCorrection = 'BH',
    signifSymbols = c('****', '***', '**', '*', ''),
    signifCutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1))
#dev.off()
```

Perform volcano plot analysis of the data
```{r volcano}
#shrink results
resWTUIlfc <- lfcShrink(dds, coef="group_WT9_vs_UI9", type="apeglm", res = resWTvsUI)
resNUIlfc <- lfcShrink(dds, coef="group_N9_vs_UI9", type="apeglm", res = resNvsUI)
#Get results for WTvs EscN
dds.N <- dds
dds.N$group <- relevel(dds.N$group, ref= "N9")
dds.N <- DESeq(dds.N)
resWTvsN <- results(dds.N, contrast = c("group","WT9", "N9"))
resWTNlfc <- lfcShrink(dds.N, coef="group_WT9_vs_N9", type="apeglm", res = resWTvsN)
#Create a mapping for ID to simplified symbol
geneids <- rownames(resWTNlfc) %>% str_replace_all("hsa-","")%>% 
  as.data.frame() %>% 
  bind_cols(rownames(resWTNlfc)) %>% 
  dplyr::select("symbol" = 1, "rowname" = `...2`)
#identify top20 genes for each comparison
id_top <- function(x) {
  x %>% 
  as.data.frame() %>% 
  arrange(padj) %>% 
  dplyr::slice(1:10) %>% 
  rownames_to_column(var = "rowname") %>% 
  left_join(geneids) }
top_list <- list("WTN" = resWTNlfc,"WTUI" = resWTUIlfc, "NUI" = resNUIlfc) %>% 
  map(id_top)
results_list <- list(resWTNlfc,resWTUIlfc,resNUIlfc)
#Replace the rownames with gene names
for (i in 1:3) {rownames(results_list [[i]]) <- geneids$symbol}
#pdf(here::here("final clean figures/Volcano plots.pdf"), width = 12, height = 10)
plotMA(resWTvsN, main = paste0("WT infection vs ", greeks("Delta"),  "EscN infection\nResults without shrinking"))
plotMA(resWTNlfc, main = paste0("WT infection vs ", greeks("Delta"),  "EscN infection\nResults after apeglm shrinking"))
plotMA(resWTvsUI, main = "WT infection vs uninfected cells \nResults without shrinking")
plotMA(resWTUIlfc, main = "WT infection vs uninfected cells \nResults after apeglm shrinking")

#Create a wrapper function for enhancedvolcano to plot all the graphs
volcano_with_labels <- function(res_obj,select_lab = "",title) {
  EnhancedVolcano(res_obj,
    lab = rownames(res_obj),
    selectLab = select_lab,
    x = 'log2FoldChange',
    y = 'pvalue',
    title = title,
    pointSize = 3,
    col = c("darkgrey","red3","dodgerblue2","forestgreen"),
    legendPosition = 'right',
    subtitle = NULL,
    pCutoff = 0.05,
    FCcutoff = 0.6,
    labSize = 6.0,
                boxedLabels = T, labCol = 'black',
               labFace = 'bold',
                drawConnectors = T, widthConnectors = 1.0,
               colConnectors = 'black')
}
titles <- list(paste0("Differentially expressed genes between WT and ",greeks("Delta"),"EscN infected cells"),
               paste0("Differentially expressed genes between WT and UI cells"),
               paste0("Differentially expressed genes between",greeks("Delta"),"EscN infected and uninfected cells"))
for (i in 1:3) {
  volcano_with_labels(results_list[[i]],title = titles [[i]]) %>% print()
  volcano_with_labels(results_list[[i]],select_lab = top_list [[i]]$gene_name,title = titles [[i]]) %>% print()
}
#dev.off()
```
Prep for EGSEA analysis to miRNA data
First get entrez ids for all miRNA names
```{r egsea-prep}
#Prepare a mart object for mapping miRNA names to 
mart <- useEnsembl(dataset = "hsapiens_gene_ensembl", biomart='ensembl')
mirs <- rownames(resWTNlfc) %>% str_remove(".*_")
annot <- getBM(attributes=c('entrezgene_id','hgnc_symbol','mirbase_id'), filters='mirbase_id', values=mirs, mart=mart)
annot_no_na_or_dup <- annot %>% filter(!is.na(entrezgene_id), !duplicated(entrezgene_id), !duplicated(mirbase_id))
```
Annotate the rows of the count matrix with entrez gene ids
```{r eggsea-prep}
#remove intercept since this is a factor. Intercept or no intercept does not make a difference
#That is what the 0+ means in the formula
design_matrix <- model.matrix(~0+pData$group) 
colnames(design_matrix) <- colnames(design_matrix) %>% str_replace("pData\\$group","") 
names_for_contr <- resultsNames(dds) [-1] %>%
  str_replace_all("group_","") %>% 
  str_replace_all("_vs_","-")
names_for_contr [3] <- "WT9-N9"
contrast_obj <- makeContrasts(contrasts = names_for_contr, levels = colnames(design_matrix))
group_obj <- as.factor(pData$group)

#Prepare results object to help inform duplicate removal (by p-values)
res <- res %>% 
  mutate(rowname = str_c(`Mature miRNA`,`Precursor miRNA`, sep = "_")) 

#finally map gene ids for the count matrix
cnt_entrez <- dataNorm@assayData$normalizedCounts %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "rowname") %>% 
  left_join(dplyr::select(res,rowname,contains("padj"), contains("log2"))) %>% 
  separate(rowname, into = c("mature","precursor"), sep = "_", remove = F) %>% 
  #Remove miRNA that have no entrez annotation
  filter(precursor %in% annot_no_na_or_dup$mirbase_id) %>% 
  group_by(precursor) %>% 
  arrange(padj.WTvsN,padj.WTvsUI,padj.NvsUI, desc(log2FoldChange.WTvsN),
          desc(log2FoldChange.WTvsUI),.by_group = T) %>% 
  dplyr::slice(1, .preserve = T) %>% 
  dplyr::select(-matches("(padj)|(log2)")) %>% 
  left_join(annot_no_na_or_dup, by = c("precursor" = "mirbase_id"))
#Create the count matrix
cnt_obj_filt <- cnt_entrez %>% 
  ungroup() %>% 
  dplyr::select(contains("9")) %>% 
  as.matrix() 
rownames(cnt_obj_filt) <- cnt_entrez$entrezgene_id
#Create a dataframe for symbol mapping
geneids_smbl <- data.frame(entrezid = cnt_entrez$entrezgene_id, symbol = cnt_entrez$hgnc_symbol)
```

Write all the necessary objects as text tables to combine with stranded library results
```{r write-tables, eval=FALSE}
write.table(cnt_obj_filt, here::here("final clean figures/mirna_count_matrix_for_egsea.txt"),
            sep = "\t", row.names = T)
write.table(counts(dds), here::here("final clean figures/mirna_count_matrix_from_dds.txt"),
            sep = "\t", row.names = T)
write.table(normalizationFactors(dds), here::here("final clean figures/mirna_norm_factors_from_dds.txt"),
            sep = "\t", row.names = T)
write_tsv(geneids_smbl, here::here("final clean figures/mirna_egsea_symbol_mapping.txt"))
```








