---
title: "Finalized script from mirDeep2 analysis of long infection"
output: 
  html_notebook: 
    fig_width: 6
    fig_height: 5
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(include = TRUE)
library(biomaRt)
library(limma)
library(greekLetters)
library(RUVSeq)
library(EDASeq)
library(DESeq2)
library(ggprism)
library(eulerr)
library(tidyverse)
library(rcartocolor)
theme_set(theme_prism(base_size = 14))
```

Define all helper functions and color choices
```{r include=TRUE}
#Set color choices from CARTOcolors
discrete_pal_choice <- "Bold"
cont_pal_choice <- "BluYl"
diverg_pal_choice <- "Geyser"
#Helper functions
calc_relative <- function(x) x/sum(x) * 100
```

Prepare the data for analysis
```{r echo=TRUE, results = 'markup', include=TRUE}
#analyzing miRNA-seq library prep for TC7 infection for 9 hours
counts <- read.table(here::here("long_infection_human_reads/miRNA_library/data/all_reads_mirDEEP2"), 
                     header = F, sep = "\t", stringsAsFactors = F)
#Next manipulate the table to remove redundant columns and only keep relevant columns (1 columns with gene names, and the rest with counts for reverse stranded library)
counts_simple <- counts %>% dplyr::select(Mature_miRNA = V1, precursor = V3, UI92 = V2, UI93 = V8, UI96 = V14, N92 = V20, N93 = V26, WT96 = V32, WT91 = V38, WT94 = V44, N95 = V50)
filter <- apply(counts_simple [,3:11],1,function(x) mean(x)>5)
counts_simple <- counts_simple [filter,]
```

Apply EDAseq normalization to the data

```{r edaseq}
#Use EDAseq to normalize the libraries
c_mat <- counts_simple %>% 
                     dplyr::select(-where(is.character)) %>% 
                     as.matrix()
rownames(c_mat) <- paste0(counts_simple$Mature_miRNA,"_",counts_simple$precursor)
pData <- data.frame(genotype=vector(mode = "logical", length = ncol(c_mat)),
                    timepoint = vector(mode = "logical", length = ncol(c_mat)),
                    samples = colnames(c_mat),
                    row.names = colnames(c_mat))
pData <- pData %>% 
  mutate(genotype = case_when(
    str_detect(samples, "WT") ~ "WT",
    str_detect(samples, "UI") ~ "UI",
    str_detect(samples, "N") ~ "N"
  ), timepoint = 9) %>%
  dplyr::select(-samples) %>% 
  mutate(graph_cols  = case_when(
    genotype == "UI" ~ carto_pal(12,discrete_pal_choice) [10],
    genotype == "WT" ~ carto_pal(12,discrete_pal_choice) [2],
    genotype == "N" ~ carto_pal(12,discrete_pal_choice) [1]
  ))

data <- newSeqExpressionSet(counts=c_mat,
                            phenoData=pData)
#Normalize the data next
dataNorm <- betweenLaneNormalization(data, which="full")
#For DE we will create this analysis with offset values
dataOffset <- betweenLaneNormalization(data,
                                       which="full",offset=TRUE)
#Plot the counts
plotPCA(data, col = pData$graph_cols, main = "Unnormalized counts", cex.axis = 1.5, cex.lab = 1.5)
plotPCA(dataOffset, col = pData$graph_cols, main = "Normalized counts", cex.axis = 1.5, cex.lab = 1.5)
plotRLE(data, col = pData$graph_cols, main = "RLE Unnormalized counts", cex.axis = 1.5)
plotRLE(dataOffset, col = pData$graph_cols, main = "RLE Normalized counts", cex.axis = 1.5)
boxplot(data, main = "Unnormalized counts", col = pData$graph_cols, cex.axis = 1.5)
boxplot(dataNorm, main = "Normalized counts", col = pData$graph_cols, cex.axis = 1.5)
```

Next step is to use normalized counts in the DESEq2 analysis of differential expression
```{r deseq}
#Prepare the DESEq2 object
pData$group <- factor(paste0(pData$genotype,pData$timepoint))
dds <- DESeqDataSetFromMatrix(countData = counts(dataOffset),
                              colData = pData,
                              design = ~ group)
dds$group <- relevel(dds$group, ref= "UI9")
normFactors <- exp(-1 * offst(dataOffset))
normFactors <- normFactors / exp(rowMeans(log(normFactors)))
normalizationFactors(dds) <- normFactors
#Perform the DESEq2 analysis
dds <- DESeq(dds)
rld <- varianceStabilizingTransformation(dds)

normal_PCA <- plotPCA(rld, returnData = T, intgroup = "group")
#Plot the PCA graphs of the normalized counts
#png(here::here("long_infection_human_reads/miRNA_library/results/final clean figures/PCA plot of normalized counts legend bottom.png"), width = 6, height = 5, units = "in", res = 1080)
ggplot(data = normal_PCA, aes(shape = pData$genotype, color = pData$genotype)) +
  scale_color_manual(labels = c("N" = paste0(greeks("Delta"),"EscN EPEC"),
                                "WT" = "Wild-type EPEC", "UI" = "Uninfected"),
                     values = c("UI" = carto_pal(12,discrete_pal_choice) [10],
                                "WT" = carto_pal(12,discrete_pal_choice) [2], 
                                "N" = carto_pal(12,discrete_pal_choice) [1])) +
  geom_point(mapping = aes(x = PC1, y = PC2), size = 5) +
  labs(title = "EDAseq-norm PCA plot", shape = "Infection", color = "Infection",
       x = "PC1: 33% variance", y = "PC2: 22% variance") +
  scale_shape_manual(labels = c("N" = paste0(greeks("Delta"),"EscN EPEC"),
                                "WT" = "Wild-type EPEC", "UI" = "Uninfected"),
                     values = c("UI" = 15, "WT" = 16, "N" = 17)) +
  theme(legend.position = "bottom", legend.title = element_text())
#dev.off()
```

Identify differentially expressed genes
```{r deseq-res-1}
resWTvsUI <- results(dds, contrast = c("group","WT9", "UI9"))
resWTvsN <- results(dds, contrast = c("group","WT9", "N9"))
resNvsUI <- results(dds, contrast = c("group","N9", "UI9"))
res <- list(resWTvsUI,resWTvsN,resNvsUI) %>% 
  map(as.data.frame) %>% 
  map(rownames_to_column, var = "rowname") %>% 
  purrr::reduce(left_join, by = "rowname") %>% 
  rename_with(function(x) str_replace(x,"$",".NvsUI"),
              .cols = !matches("\\.[xy]")) %>% 
  rename_with(function(x) ifelse(str_detect(x,"\\.x"),
                                 str_replace(x,"\\.x",".WTvsUI"),
                                 str_replace(x,"\\.y",".WTvsN")),
              .cols = matches("\\.[xy]")) %>% 
  rename_with(function(x) "Gene ID",1) %>% 
  separate(`Gene ID`, into = c("Mature miRNA", "Precursor miRNA"), sep = "_")
walk(list(resWTvsUI,resWTvsN,resNvsUI),summary)
#write.csv(res, here::here("long_infection_human_reads/miRNA_library/results/final clean figures/tables/normalized results.csv"), row.names = F)
```
Next step is to determine overwhelming presence of miRNA (some can make up 50% of the data)
```{r abund-mirna, fig.width=8, fig.height=6}
calc_relative <- function(x) x/sum(x)*100
rel_abund <- counts(dds) %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "rowname") %>% 
  mutate(across(!where(is.character), calc_relative)) %>%  
  arrange(desc(UI92),desc(UI93)) %>% 
  mutate(top_names = c(rowname [1:5],rep("other",nrow(.)-5))) %>% 
  pivot_longer(-c("rowname","top_names"), names_to = "mapping", values_to = "numbers")
#remove the highest abundant miRNA to see how the analysis falls
rel_abund_no_high <- counts(dds) %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "rowname") %>% 
  filter(rowname != "hsa-miR-192-5p_hsa-mir-192" & rowname != "hsa-miR-215-5p_hsa-mir-215") %>%
  mutate(across(!where(is.character), calc_relative)) %>%  
  arrange(desc(UI92),desc(UI93)) %>% 
  mutate(top_names = c(rowname [1:5],rep("other",nrow(.)-5))) %>% 
  pivot_longer(-c("rowname","top_names"), names_to = "mapping", values_to = "numbers")
#pdf(here::here("long_infection_human_reads/miRNA_library/results/final clean figures/miRNA mapping stats by miRNA identity1.pdf"),width = 8,height = 6)
rel_list <- list(rel_abund,rel_abund_no_high)
rel_point <- rel_abund %>% 
  group_by(top_names, mapping) %>% 
  summarise(numbers = mean(numbers)) %>% 
  mutate(condition = case_when(
    str_detect(mapping,"WT") ~ "WT",
    str_detect(mapping,"N") ~ "N",
    str_detect(mapping,"UI") ~ "UI",
  ))
for (i in 1:2) {
  p <- ggplot(rel_list [[i]], aes(x=mapping, y=numbers, fill = fct_reorder(top_names,numbers, .desc = T))) +
    geom_bar(stat = "identity") + scale_fill_carto_d(name = "Aligned to miRNA",palette = "Bold") +
    labs(x = "", y = "Proportion of counts") +
    theme(axis.text.x = element_text(angle = 75, vjust = 0.75, hjust = 0.6))
  print(p)
}

ggplot(rel_point, aes(x = mapping, y = fct_reorder(top_names,numbers, .desc = T), size = numbers, fill = condition)) + 
  geom_point(alpha = 0.75, shape = 21) + 
  scale_size_continuous(range = c(3,10), breaks = c(2,5,8,10)) + 
  labs( x= "", y = "", size = "Relative Abundance (%)", fill = "Condition")  + 
  scale_fill_manual(name = "Infection status",
                    labels = c("N" = bquote(Delta*"EscN EPEC"),
                               "WT" = "Wild-type EPEC", "UI" = "Uninfected"),
                    values = c("UI" = carto_pal(12,discrete_pal_choice) [10],
                               "WT" = carto_pal(12,discrete_pal_choice) [2], 
                               "N" = carto_pal(12,discrete_pal_choice) [1])) +
  theme_prism(axis_text_angle = 45, base_size = 13) +
  theme(legend.title = element_text())
#dev.off()
```

Remove the two highest abundant miRNA from analysis and repeat analysis
```{r deseq-1}
data.n <- newSeqExpressionSet(counts=c_mat [rownames(c_mat) != "hsa-miR-192-5p_hsa-mir-192" 
                                          & rownames(c_mat) != "hsa-miR-215-5p_hsa-mir-215",],
                            phenoData=pData)
#Normalize the data next
dataNorm.n <- betweenLaneNormalization(data.n, which="full") 
#Plot the counts 
#pdf(here::here("final clean figures/Normalization effect on the library.pdf"), width = 12, height = 10)
boxplot(data.n, main = "Unnormalized counts", col = "dodgerblue3")
boxplot(dataNorm.n, main = "Normalized counts", col = "dodgerblue3")
#dev.off()
#For DE we will create this analysis with offset values
dataOffset.n <- betweenLaneNormalization(data.n,
                                       which="full",offset=TRUE)
#Prepare the DESEq2 object and apply EDASeq normalization factors
pData$group <- factor(paste0(pData$genotype,pData$timepoint))
dds.new <- DESeqDataSetFromMatrix(countData = counts(dataOffset.n),
                              colData = pData,
                              design = ~ group)
dds.new$group <- relevel(dds.new$group, ref= "UI9")
normFactors <- exp(-1 * offst(dataOffset))
normFactors <- normFactors / exp(rowMeans(log(normFactors)))
normalizationFactors(dds) <- normFactors
#Perform the DESEq2 analysis
dds.new <- DESeq(dds.new)
```
Identify differentially expressed genes
```{r deseq-res}
resWTvsUI.new <- results(dds.new, contrast = c("group","WT9", "UI9"))
resWTvsN.new <- results(dds.new, contrast = c("group","WT9", "N9"))
resNvsUI.new <- results(dds.new, contrast = c("group","N9", "UI9"))
summary(resWTvsUI.new)
summary(resWTvsN.new)
#Removing top 2 miRNA does not really affect differential expression analysis, so keep them in
```

Perform PCA tools analysis of the data
```{r pcatools}
cnts <- assay(rld)
#Confirm that that the mapping is correct
rownames(cnts) <- rownames(cnts) %>% str_replace_all("hsa-","")
p <- pca(cnts,colData(dds), removeVar = 0.1)
pca_coord_meta <- p$rotated %>% 
  merge(as.data.frame(colData(rld2.nobatch)), by = 0) %>% 
  mutate(group = fct_relevel(group, "WT9"))
#Create the lists to store information on the linear model performance
fit_list <- set_names(vector(mode = "list", length = 9),
                              nm = str_c("PC",1:9))
signi_list <- set_names(vector(mode = "list", length = 9),
                              nm = str_c("PC",1:9))
#Run linear regression for each principal component
for (PC in seq_along(fit_list)) {
  frm <- as.formula(str_c("PC",PC, " ~ group"))
  fit_list [[PC]] <- lm(frm, data = pca_coord_meta)
  sum_of_fit <- summary(fit_list [[PC]])
  signi_list [[PC]] <- sum_of_fit$coefficients %>% 
    as.data.frame() %>% 
    mutate(R_squared = sum_of_fit$r.squared)
}
#Simplify the lists
#Summarise liner model performance and plot the results
signi_list_reduced <- signi_list %>% 
  map2(names(signi_list), ~mutate(.x, PC = .y)) %>% 
  map(rownames_to_column, var = "coef") %>% 
  purrr::reduce(bind_rows) %>% 
  filter(coef != "(Intercept)") %>% 
  dplyr::rename(p = `Pr(>|t|)`) %>%  
  mutate(signi = case_when(
    p < 0.05 ~ "yes",
    TRUE ~ "no"
  ))
#Plot the results of regression
ggplot(signi_list_reduced) +
  geom_tile(aes(y=coef,x=PC, fill = p)) +
  geom_text(data = filter(signi_list_reduced, signi == "yes"), aes(y=coef,x=PC), 
            label = "*", color = "white", size = 13) +
  labs(x = element_blank(), y = element_blank(), title = "Linear regression analysis\nidentifying relationships between metadata and PC") +
  scale_y_discrete(labels = c(groupUI9 = "Effect of Infection",groupN9 = "Effect of T3SS")) +
  scale_fill_carto_c(palette = "BluYl", direction = -1, name = "p-value") +
  theme_minimal(base_size = 14) +
  theme(panel.border = element_rect(fill = NA, color = "black"),
        axis.ticks.x = element_line(color = "black"),
        axis.ticks.y = element_line(color = "black"),
        panel.grid.major = element_blank(),
        axis.text = element_text(color = "black"),
        title = element_text(size = 8),
        legend.title = element_text(size = 14))
#ggsave(here::here("long_infection_human_reads/miRNA_library/results/final clean figures/Linear regression of metadata and PCA.png"), width = 6,height = 2.5, dpi = 1080)
```


```{r pcatools-loadings-plot, fig.width=9, fig.height=6}
plotloadings(p,  rangeRetain = 0.01,
    labSize = 2.5, col = carto_pal(3,diverg_pal_choice) ,
    caption = "Top 1% variables") + theme(legend.title = element_text())
#ggsave(here::here("long_infection_human_reads/miRNA_library/results/final clean figures/PCA plot loadins.png"), width = 9, height = 6, dpi = 1080)
```

Perform volcano plot analysis of the data
```{r volcano}
#shrink results
resWTUIlfc <- lfcShrink(dds, coef="group_WT9_vs_UI9", type="apeglm", res = resWTvsUI)
resNUIlfc <- lfcShrink(dds, coef="group_N9_vs_UI9", type="apeglm", res = resNvsUI)
#Get results for WTvs EscN
dds.N <- dds
dds.N$group <- relevel(dds.N$group, ref= "N9")
dds.N <- DESeq(dds.N)
resWTvsN <- results(dds.N, contrast = c("group","WT9", "N9"))
resWTNlfc <- lfcShrink(dds.N, coef="group_WT9_vs_N9", type="apeglm", res = resWTvsN)
#Create a mapping for ID to simplified symbol
geneids <- rownames(resWTNlfc) %>% str_replace_all("hsa-","")%>% 
  as.data.frame() %>% 
  bind_cols(rownames(resWTNlfc)) %>% 
  dplyr::select("symbol" = 1, "rowname" = `...2`)

results_list <- list(resWTNlfc,resWTUIlfc,resNUIlfc)
#Replace the rownames with gene names
for (i in 1:3) {rownames(results_list [[i]]) <- geneids$symbol}
#Prepare the results list for plotting
res_df_list <- results_list %>% 
  map(as.data.frame) %>% 
  map(rownames_to_column, var = "gene") %>% 
  map(mutate, signi = if_else(padj < 0.05 & abs(log2FoldChange) >= 0.5, "yes","no"))

#Create the function for making a nice volcano plot
volcano_ggplot <- function(res_df, title = "", to_label = F, n_lab = 10) {
  p_empty <- ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj), fill = signi, color = signi)) +
    geom_hline(yintercept = -log10(0.05), lty = 2, lwd = 0.8, color = "grey30") +
    geom_vline(xintercept = -0.5, lty = 2, lwd = 0.8, color = "grey30") +
    geom_vline(xintercept = 0.5, lty = 2, lwd = 0.8, color = "grey30") +
    geom_point(size = 4, shape = 21, alpha = 0.7) +
    scale_color_manual(values = c("darkgrey", carto_pal(7,diverg_pal_choice) [7])) +
    scale_fill_manual(values = c("darkgrey", carto_pal(7,diverg_pal_choice) [7])) +
    guides(color = 'none', fill = 'none') +
    labs(x = bquote(~Log[2] ~ "fold change"),
         y = bquote(~-Log[10] ~ italic(P)),
         title = title) +
    theme_prism(base_size = 24) +
    theme(title = element_text(size = 10),
          axis.title = element_text(size = 24))
  if(to_label) {
    res_df_top <- res_df %>% 
      filter(signi == "yes") %>% 
      arrange(desc(abs(log2FoldChange))) %>% 
      slice_head(n=n_lab)
    p_labelled <- p_empty + 
      geom_label_repel(data = res_df_top, aes(x = log2FoldChange, y = -log10(padj), label = gene),
                       size = 4, max.overlaps = Inf, color = "black", fill = "white")
    return(p_labelled)
  } else {
    return(p_empty)
  }
}
#Print the volcano plots - each should get its own png file
volcano_list <- map2(res_df_list, 
                     list(bquote("WT infection vs"~Delta*"EscN infection"),
                          "WT infection vs Uninfected cells",
                          bquote(Delta*"EscN infection vs Uninfected cells")),
                     volcano_ggplot)
#Repeat the same with adding top 10 DE genes
volcano_list_labelled <- map2(res_df_list, 
                     list(bquote("WT infection vs"~Delta*"EscN infection"),
                          "WT infection vs Uninfected cells",
                          bquote(Delta*"EscN infection vs Uninfected cells")),
                     volcano_ggplot, to_label = T, n_lab = 10)
#pdf(here::here("long_infection_human_reads/miRNA_library/results/final clean figures/Volcano plots.pdf"), width = 8, height = 6)
walk(volcano_list, print)
walk(volcano_list_labelled, print)
#dev.off()



#pdf(here::here("long_infection_human_reads/miRNA_library/results/final clean figures/MA plots.pdf"), width = 6, height = 5)
DESeq2::plotMA(resWTvsN, main = paste0("WT infection vs ", greeks("Delta"),  "EscN infection\nResults without shrinking"))
DESeq2::plotMA(resWTNlfc, main = paste0("WT infection vs ", greeks("Delta"),  "EscN infection\nResults after apeglm shrinking"))
DESeq2::plotMA(resWTvsUI, main = "WT infection vs Uninfected cells \nResults without shrinking")
DESeq2::plotMA(resWTUIlfc, main = "WT infection vs Uninfected cells \nResults after apeglm shrinking")
#dev.off()
```


Write all the necessary objects as text tables to combine with stranded library results
```{r write-tables, eval=FALSE}
write.table(counts(dds), here::here("long_infection_human_reads/miRNA_library/results/final clean figures/tables/mirna_count_matrix_from_dds.txt"),
            sep = "\t", row.names = T)
write.table(normalizationFactors(dds), here::here("long_infection_human_reads/miRNA_library/results/final clean figures/tables/mirna_norm_factors_from_dds.txt"),
            sep = "\t", row.names = T)
write.table(assay(rld), here::here("long_infection_human_reads/miRNA_library/results/final clean figures/tables/mirna_rld_count_matrix_from_dds.txt"),
            sep = "\t", row.names = T)
```








